In [4]:
from google.colab import drive
import os
import pandas as pd
from scipy import stats
import numpy as np

In [5]:
drive.mount('/content/drive')

DATA_DIR = '/content/drive/MyDrive/data'

# 3. Automatically change the working directory so relative paths work
if os.path.exists(DATA_DIR):
    os.chdir(DATA_DIR)
    print(f"Successfully connected! Current working directory: {os.getcwd()}")
else:
    print(f"Error: The folder '{DATA_DIR}' could not be found.")
    print("Make sure to add a shortcut of the 'data' folder to your 'My Drive'.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Successfully connected! Current working directory: /content/drive/.shortcut-targets-by-id/1_ib6SYLLQ6Asjybns0GSlgBDwFQrnmkj/data


### Uploading Data and Addressing Missing Values and Outliers

## Data Cleaning Methodology

The datasets are cleaned by addressing missing values and potential outliers.

### Missing values

Missing values are handled according to the meaning of the variable rather than
using one imputation method for every column.

The following rules are used:

- Sentinel values representing missing data are converted to `NaN`.
- Rows missing an essential measurement are removed when the observation cannot
  be used for the analysis.
- Short gaps in sequential environmental measurements may be filled using
  linear interpolation when nearby observations provide a reasonable estimate.
- Missing categorical metadata is represented as `"Unknown"` when appropriate.
- Measurement or metadata values that cannot be reliably inferred are left as
  `NaN`.
- Median imputation is used only when a numerical variable can reasonably be
  represented by a typical value and interpolation is not appropriate.

### Outliers

The Interquartile Range (IQR) method is used consistently for outlier screening.

For each measurement:

- IQR = Q3 - Q1
- Lower bound = Q1 - 1.5 × IQR
- Upper bound = Q3 + 1.5 × IQR

Values outside these bounds are considered statistical outliers. However,
statistical outliers are not automatically removed. They are investigated using
their units, sampling context, location, quality information, and physical
plausibility.

Values are removed only when there is evidence that they are clearly erroneous
or physically implausible.

In [6]:
# Load all datasets
bradentonFL = pd.read_csv("bradentonFL_tempData.csv")
floridaCyanotoxins = pd.read_csv("florida_cyanotoxins_2015-01-01_to_2026-06-17.csv")
habsos = pd.read_csv("habsos_cellcounts.csv")
motemarine = pd.read_csv("MoteMarine_BottomTemp.csv")
nutrientsFL = pd.read_csv("NutrientsFL_2006_2025.csv")
sarasota = pd.read_csv("Sarasota_Wind&Temp.csv")

/tmp/ipykernel_1019/3135485723.py:3: DtypeWarning: Columns (48) have mixed types. Specify dtype option on import or set low_memory=False.
  floridaCyanotoxins = pd.read_csv("florida_cyanotoxins_2015-01-01_to_2026-06-17.csv")
/tmp/ipykernel_1019/3135485723.py:4: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  habsos = pd.read_csv("habsos_cellcounts.csv")


## bradentonFL_tempData.csv

In [7]:
print("Shape:", bradentonFL.shape)

Shape: (4018, 8)


In [8]:
bradentonFL.head()

,COOPID,YEAR,MONTH,DAY,PRECIPITATION,MAX TEMP,MIN TEMP,MEAN TEMP
0,80945,2015,1,1,0.0,74.0,60.0,
1,80945,2015,1,2,0.0,78.0,64.0,
2,80945,2015,1,3,0.0,85.0,70.0,
3,80945,2015,1,4,0.0,83.0,69.0,
4,80945,2015,1,5,0.0,75.0,61.0,


In [9]:
# Check missing values before cleaning.
print("Missing values:")
print(bradentonFL.isna().sum())

Missing values:
COOPID            0
 YEAR             0
 MONTH            0
 DAY              0
 PRECIPITATION    0
 MAX TEMP         0
 MIN TEMP         0
 MEAN TEMP        0
dtype: int64


In [10]:
print("\nColumn names:")
print(bradentonFL.columns)


Column names:
Index(['COOPID', ' YEAR', ' MONTH', ' DAY', ' PRECIPITATION', ' MAX TEMP',
       ' MIN TEMP', ' MEAN TEMP'],
      dtype='object')


In [11]:
# Rename columns for easier reference
bradentonFL.rename(
    columns={
        'COOPID': "coop_id",
        ' YEAR': 'year',
        ' MONTH': 'month',
        ' DAY': 'day',
        ' PRECIPITATION': 'precipitation',
        ' MAX TEMP': 'max_temp',
        ' MIN TEMP': 'min_temp',
        ' MEAN TEMP': 'mean_temp'
    },
    inplace=True
)

In [12]:
# Remove mean temperature because it will not needed for the analysis
bradentonFL.drop(columns='mean_temp', inplace=True)

In [13]:
bradentonFL.describe()

,coop_id,year,month,day,precipitation,max_temp,min_temp
count,4018.0,4018.00000,4018.000000,4018.000000,4018.000000,4018.000000,4018.000000
mean,80945.0,2020.00000,6.522648,15.730463,0.028726,84.121379,66.079841
std,0.0,3.16275,3.449210,8.801536,3.565303,11.761406,12.512208
min,80945.0,2015.00000,1.000000,1.000000,-99.990000,-99.900000,-99.900000
25%,80945.0,2017.00000,4.000000,8.000000,0.000000,80.000000,60.000000
50%,80945.0,2020.00000,7.000000,16.000000,0.000000,86.000000,69.000000
75%,80945.0,2023.00000,10.000000,23.000000,0.040000,91.000000,75.000000
max,80945.0,2025.00000,12.000000,31.000000,11.650000,101.000000,83.000000


### Bradenton Missing Values

The precipitation and temperature columns contain sentinel values such as
`-99.99` and `-99.9`. These values represent missing observations rather than
actual environmental measurements, so they are converted to `NaN`.

Missing weather measurements are not median-imputed because doing so would
create artificial weather observations for days when measurements were not
recorded.

In [14]:
# Check for sentinel values representing missing temperature or precipitation data
bradentonFL[
    ["precipitation", "max_temp", "min_temp"]
].isin([-99.99, -99.9]).sum()

,0
precipitation,5
max_temp,7
min_temp,8


In [15]:
# Replace sentinel values with NaN so they are treated as missing values
bradentonFL["precipitation"] = bradentonFL["precipitation"].replace([-99.99, -99.9], np.nan)
bradentonFL["max_temp"] = bradentonFL["max_temp"].replace(-99.9, np.nan)
bradentonFL["min_temp"] = bradentonFL["min_temp"].replace(-99.9, np.nan)

In [16]:
# Check for any remaining negative precipitation values
print(
    bradentonFL[
        ["precipitation", "max_temp", "min_temp"]
    ].isin([-99.99, -99.9]).sum()
)

print("\nRemaining missing values:")
print(bradentonFL.isna().sum())

precipitation    0
max_temp         0
min_temp         0
dtype: int64

Remaining missing values:
coop_id          0
year             0
month            0
day              0
precipitation    5
max_temp         7
min_temp         8
dtype: int64


In [17]:
# Check summary statistics after replacing sentinel values.
bradentonFL.describe()

,coop_id,year,month,day,precipitation,max_temp,min_temp
count,4018.0,4018.00000,4018.000000,4018.000000,4013.000000,4011.000000,4010.000000
mean,80945.0,2020.00000,6.522648,15.730463,0.153277,84.442533,66.410973
std,0.0,3.16275,3.449210,8.801536,0.507824,8.908142,10.088764
min,80945.0,2015.00000,1.000000,1.000000,0.000000,45.000000,29.000000
25%,80945.0,2017.00000,4.000000,8.000000,0.000000,80.000000,60.000000
50%,80945.0,2020.00000,7.000000,16.000000,0.000000,86.000000,69.000000
75%,80945.0,2023.00000,10.000000,23.000000,0.040000,91.000000,75.000000
max,80945.0,2025.00000,12.000000,31.000000,11.650000,101.000000,83.000000


In [18]:
# Check for any remaining negative precipitation values.
bradentonFL[
    bradentonFL["precipitation"] < 0
][
    ["year", "month", "day", "precipitation"]
]

,year,month,day,precipitation


In [19]:
# Check whether minimum temperature exceeds maximum temperature.
print(
    "Rows where min_temp > max_temp:",
    (bradentonFL["min_temp"] > bradentonFL["max_temp"]).sum()
)

Rows where min_temp > max_temp: 0


### Bradenton Outlier Detection

IQR is applied to precipitation, maximum temperature, and minimum temperature.
The date variables and station identifier are not treated as environmental
measurement outliers.

IQR outliers are investigated rather than automatically deleted because extreme
weather conditions can be legitimate observations.

In [20]:
# Variables used for IQR outlier screening.
measurement_cols = [
    "precipitation",
    "max_temp",
    "min_temp"
]

# Calculate Q1, Q3, and IQR.
Q1 = bradentonFL[measurement_cols].quantile(0.25)
Q3 = bradentonFL[measurement_cols].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Lower bounds:")
print(lower_bound)

print("\nUpper bounds:")
print(upper_bound)

# Identify IQR outliers.
outliers = (
    (bradentonFL[measurement_cols] < lower_bound) |
    (bradentonFL[measurement_cols] > upper_bound)
)

print("\nNumber of IQR outliers:")
print(outliers.sum())

Lower bounds:
precipitation    -0.06
max_temp         63.50
min_temp         37.50
dtype: float64

Upper bounds:
precipitation      0.1
max_temp         107.5
min_temp          97.5
dtype: float64

Number of IQR outliers:
precipitation    810
max_temp         117
min_temp          17
dtype: int64


In [21]:
# Inspect precipitation outliers.
display(
    bradentonFL[outliers["precipitation"]][
        ["year", "month", "day", "precipitation"]
    ].sort_values(
        "precipitation",
        ascending=False
    ).head(20)
)

# Inspect maximum-temperature outliers.
display(
    bradentonFL[outliers["max_temp"]][
        ["year", "month", "day", "max_temp"]
    ].sort_values(
        "max_temp",
        ascending=False
    ).head(20)
)

# Inspect minimum-temperature outliers.
display(
    bradentonFL[outliers["min_temp"]][
        ["year", "month", "day", "min_temp"]
    ].sort_values(
        "min_temp"
    ).head(20)
)

,year,month,day,precipitation
968,2017,8,26,11.65
2827,2022,9,28,7.09
3503,2024,8,4,7.00
489,2016,5,4,6.74
608,2016,8,31,6.71
3504,2024,8,5,6.30
1980,2020,6,3,6.03
2141,2020,11,11,5.80
3569,2024,10,9,4.62
1449,2018,12,20,4.30


,year,month,day,max_temp
40,2015,2,10,63.0
367,2016,1,3,63.0
708,2016,12,9,63.0
420,2016,2,25,63.0
421,2016,2,26,63.0
1110,2018,1,15,63.0
1102,2018,1,7,63.0
1176,2018,3,22,63.0
1125,2018,1,30,63.0
1096,2018,1,1,63.0


,year,month,day,min_temp
1113,2018,1,18,29.0
50,2015,2,20,34.0
2915,2022,12,25,34.0
1075,2017,12,11,34.0
2914,2022,12,24,35.0
2587,2022,1,31,35.0
1074,2017,12,10,35.0
389,2016,1,25,36.0
1100,2018,1,5,36.0
1099,2018,1,4,36.0


### Bradenton Weather Outlier Assessment

The IQR method identified statistical outliers in precipitation, maximum
temperature, and minimum temperature. These values were investigated to
determine whether they represented data errors or potentially valid weather
observations.

The identified outliers were retained because no negative precipitation
values or inconsistencies between minimum and maximum temperatures were found.
The observed extreme weather measurements may represent legitimate weather
conditions rather than erroneous data.

## florida_cyanotoxins_2015-01-01_to_2026-06-17.csv

In [22]:
print("Shape:", floridaCyanotoxins.shape)

Shape: (98884, 65)


In [23]:
floridaCyanotoxins.head()

,Org_Identifier,Org_FormalName,Project_Identifier,Location_Identifier,Location_Name,Location_Type,Location_State,Location_HUCEightDigitCode,Location_HUCTwelveDigitCode,Location_TribalLand,...,Result_DetectionLimitProfileDow,Result_LabSamplePrepProfileDown,ProviderName,Result_CharacteristicComparable,Result_CharacteristicGroup,Org_Type,LastChangeDate,USGSpcode,ObjectId,Value_control
0,NARS_WQX,EPA National Aquatic Resources Survey (NARS),"[""NARS_NCCA2015""]",NARS_WQX-NCA_FL-10246,NCA_FL-10246,Ocean,Florida,NaN,NaN,NaN,...,https://www.waterqualitydata.us/data/providers...,NaN,STORET,NaN,"Cyanotoxins, Phytotoxins",Federal/US Government,Fri Jan 31 08:59:46 GMT 2025,NaN,23431,A
1,NARS_WQX,EPA National Aquatic Resources Survey (NARS),"[""NARS_NCCA2015""]",NARS_WQX-NCA_FL-10245,NCA_FL-10245,Ocean,Florida,NaN,NaN,NaN,...,https://www.waterqualitydata.us/data/providers...,NaN,STORET,NaN,"Cyanotoxins, Phytotoxins",Federal/US Government,Fri Jan 31 08:59:46 GMT 2025,NaN,22625,A
2,NARS_WQX,EPA National Aquatic Resources Survey (NARS),"[""NARS_NCCA2015""]",NARS_WQX-NCA_FL-10297,NCA_FL-10297,Ocean,Florida,NaN,NaN,NaN,...,https://www.waterqualitydata.us/data/providers...,NaN,STORET,NaN,"Cyanotoxins, Phytotoxins",Federal/US Government,Fri Jan 31 08:59:46 GMT 2025,NaN,22713,A
3,NARS_WQX,EPA National Aquatic Resources Survey (NARS),"[""NARS_NCCA2015""]",NARS_WQX-NCA_FL-10299,NCA_FL-10299,Ocean,Florida,NaN,NaN,NaN,...,https://www.waterqualitydata.us/data/providers...,NaN,STORET,NaN,"Cyanotoxins, Phytotoxins",Federal/US Government,Fri Jan 31 08:59:46 GMT 2025,NaN,25340,A
4,NARS_WQX,EPA National Aquatic Resources Survey (NARS),"[""NARS_NCCA2015""]",NARS_WQX-NCA_FL-10248,NCA_FL-10248,Ocean,Florida,NaN,NaN,NaN,...,https://www.waterqualitydata.us/data/providers...,NaN,STORET,NaN,"Cyanotoxins, Phytotoxins",Federal/US Government,Fri Jan 31 08:59:46 GMT 2025,NaN,25355,A


In [24]:
# Check missing values.
print(
    floridaCyanotoxins.isna().sum()[
        floridaCyanotoxins.isna().sum() > 0
    ]
)

Location_HUCEightDigitCode           188
Location_HUCTwelveDigitCode          308
Location_TribalLand                98884
Activity_MediaSubdivision            236
ActivityBiological_AssemblageSa    98884
Activity_StartTime                   220
Activity_StartTimeZone               220
Activity_DepthHeightMeasure          228
Activity_DepthHeightMeasureUnit      228
Result_ResultDetectionCondition    98884
Result_CharacteristicUserSuppli    98884
Result_MethodSpeciation            98884
Result_SampleFraction                220
ResultBiological_Intent            98884
ResultBiological_Taxon             98884
ResultDepthHeight_Measure          98884
ResultDepthHeight_MeasureUnit      98884
Reslt_Measre                         131
Result_MeasureUnit                    60
Result_MeasureQualifierCode        98884
Result_StatisticalBase             98884
DataQuality_ResultComment           1772
DetectionLimit_TypeA                  60
DetectionLimit_MeasureA               60
DetectionLimit_M

In [25]:
# Remove columns that contain no data.
floridaCyanotoxins = floridaCyanotoxins.dropna(
    axis=1,
    how="all"
)

### Cyanotoxin Missing Values

`Reslt_Measre` is the primary measurement used for the analysis. Rows without
this measurement cannot contribute to the cyanotoxin analysis and are therefore
removed.

Other numerical fields, such as depth or geographic coordinates, are not
automatically median-imputed because replacing them with a typical value could
create artificial sampling information.

Categorical fields with missing values are represented as `"Unknown"`.

In [26]:
floridaCyanotoxins["Reslt_Measre"].describe()

,Reslt_Measre
count,98753.000000
mean,0.449555
std,7.271026
min,0.060000
25%,0.100000
50%,0.250000
75%,0.250000
max,780.000000


In [27]:
# Remove rows where the primary cyanotoxin measurement is missing.
floridaCyanotoxins = floridaCyanotoxins.dropna(
    subset=["Reslt_Measre"]
)

In [28]:
# Confirm that all remaining rows have a cyanotoxin measurement.
floridaCyanotoxins["Reslt_Measre"].isna().sum()

np.int64(0)

In [29]:
floridaCyanotoxins[
    floridaCyanotoxins["Reslt_Measre"] < 0
]

,Org_Identifier,Org_FormalName,Project_Identifier,Location_Identifier,Location_Name,Location_Type,Location_State,Location_HUCEightDigitCode,Location_HUCTwelveDigitCode,Location_Latitude,...,LabInfo_AnalysisStartDate,LabInfo_AnalysisStartTime,LabInfo_AnalysisStartTimeZone,Result_DetectionLimitProfileDow,ProviderName,Result_CharacteristicGroup,Org_Type,LastChangeDate,ObjectId,Value_control


In [30]:
# Separate numerical and categorical columns for missing-value treatment
numerical_cols = floridaCyanotoxins.select_dtypes(include=["number"]).columns
categorical_cols = floridaCyanotoxins.select_dtypes(include=["object"]).columns

In [31]:
# Fill missing categorical values with "Unknown"
floridaCyanotoxins[categorical_cols] = floridaCyanotoxins[
    categorical_cols
].fillna("Unknown")

In [32]:
print("Remaining missing values:")
print(
    floridaCyanotoxins.isna().sum()[
        floridaCyanotoxins.isna().sum() > 0
    ]
)

Remaining missing values:
Location_HUCEightDigitCode      93
Location_HUCTwelveDigitCode    177
Activity_DepthHeightMeasure     97
DetectionLimit_MeasureA         60
dtype: int64


### Cyanotoxin Outlier Detection

IQR is applied to `Reslt_Measre`, the primary cyanotoxin measurement.

Because cyanotoxin concentrations can be strongly right-skewed, unusually high
measurements are expected to occur. Therefore, IQR outliers are investigated
before any decision to remove them.

In [33]:
# Examine the original distribution before removing any outliers.
floridaCyanotoxins["Reslt_Measre"].describe()

,Reslt_Measre
count,98753.000000
mean,0.449555
std,7.271026
min,0.060000
25%,0.100000
50%,0.250000
75%,0.250000
max,780.000000


In [34]:
# Calculate IQR bounds for each cyanotoxin characteristic.
iqr_by_characteristic = (
    floridaCyanotoxins
    .groupby("Result_Characteristic")["Reslt_Measre"]
    .agg(
        Q1=lambda x: x.quantile(0.25),
        Q3=lambda x: x.quantile(0.75)
    )
)

iqr_by_characteristic["IQR"] = (
    iqr_by_characteristic["Q3"] -
    iqr_by_characteristic["Q1"]
)

iqr_by_characteristic["Lower_Bound"] = (
    iqr_by_characteristic["Q1"] -
    1.5 * iqr_by_characteristic["IQR"]
)

iqr_by_characteristic["Upper_Bound"] = (
    iqr_by_characteristic["Q3"] +
    1.5 * iqr_by_characteristic["IQR"]
)

iqr_by_characteristic

,Q1,Q3,IQR,Lower_Bound,Upper_Bound
Result_Characteristic,,,,,
Anatoxin-A,0.25000,0.250,0.00000,0.250000,0.250000
Cylindrospermopsin,0.10000,0.250,0.15000,-0.125000,0.475000
Microcystin,0.18400,0.194,0.01000,0.169000,0.209000
Microcystin HilR,0.25000,0.250,0.00000,0.250000,0.250000
Microcystin HtyR,0.25000,0.250,0.00000,0.250000,0.250000
Microcystin LA,0.10000,0.250,0.15000,-0.125000,0.475000
Microcystin LF,0.10000,0.250,0.15000,-0.125000,0.475000
Microcystin LR,0.25000,0.380,0.13000,0.055000,0.575000
"Microcystin LR, Desmethyl",0.25000,0.250,0.00000,0.250000,0.250000


In [35]:
# Count IQR outliers for each cyanotoxin characteristic.
outlier_counts = []

# Process each cyanotoxin characteristic separately.
for characteristic, row in iqr_by_characteristic.iterrows():

    # Get the measurements for the current characteristic.
    values = floridaCyanotoxins.loc[
        floridaCyanotoxins["Result_Characteristic"] == characteristic,
        "Reslt_Measre"
    ]

    # Identify measurements outside the characteristic-specific IQR bounds.
    outliers = values[
        (values < row["Lower_Bound"]) |
        (values > row["Upper_Bound"])
    ]

    # Store the number of statistical outliers.
    outlier_counts.append({
        "Result_Characteristic": characteristic,
        "Outlier_Count": len(outliers)
    })

# Convert the results into a DataFrame.
outlier_counts = pd.DataFrame(outlier_counts)

# Display characteristics with the most outliers first.
outlier_counts.sort_values(
    "Outlier_Count",
    ascending=False
)

,Result_Characteristic,Outlier_Count
7,Microcystin LR,1398
5,Microcystin LA,305
8,"Microcystin LR, Desmethyl",257
11,Microcystin RR,177
18,Saxitoxin,154
1,Cylindrospermopsin,145
13,Microcystin YR,98
3,Microcystin HilR,90
17,Nodularin,87
12,Microcystin WR,86


In [36]:
# Display the largest cyanotoxin measurements for contextual inspection.
extreme_cyanotoxins = (
    floridaCyanotoxins
    .loc[
        floridaCyanotoxins["Reslt_Measre"] > 100,
        [
            "Location_Name",
            "Location_State",
            "Result_Characteristic",
            "Reslt_Measre",
            "Result_MeasureUnit",
            "Result_MeasureStatusIdentifier",
            "Result_MeasureValueType"
        ]
    ]
    .sort_values(
        "Reslt_Measre",
        ascending=False
    )
)

extreme_cyanotoxins

,Location_Name,Location_State,Result_Characteristic,Reslt_Measre,Result_MeasureUnit,Result_MeasureStatusIdentifier,Result_MeasureValueType
19925,Lake Okeechobee - Pahokee Marina,Florida,Microcystin LR,780.0,ug/L,Final,Actual
11013,L004,Florida,Microcystin LR,760.0,ug/L,Final,Actual
76064,C44 Canal - Timer Powers Park Boat Ramp,Florida,Microcystin LR,760.0,ug/L,Final,Actual
56531,Moody Lake - SE,Florida,Microcystin LR,750.0,ug/L,Final,Actual
362,S352,Florida,Microcystin LR,730.0,ug/L,Final,Actual
17648,Harbor Isles - Southern Lobe,Florida,Microcystin LR,600.0,ug/L,Final,Actual
17671,Harbor Isles - Southern Lobe,Florida,Microcystin LR,600.0,ug/L,Final,Actual
2367,C44S80 St. Lucie Lock,Florida,Microcystin LR,490.0,ug/L,Final,Actual
1049,S79 Upstream of Lock,Florida,Microcystin LR,450.0,ug/L,Final,Actual
34048,Boat Ramp,Florida,Microcystin LR,410.0,ug/L,Final,Actual


In [37]:
# Check the highest cyanotoxin measurements.
floridaCyanotoxins.nlargest(
    20,
    "Reslt_Measre"
)[
    [
        "Result_Characteristic",
        "Reslt_Measre",
        "Result_MeasureUnit",
        "Result_MeasureStatusIdentifier",
        "Result_MeasureValueType"
    ]
]

,Result_Characteristic,Reslt_Measre,Result_MeasureUnit,Result_MeasureStatusIdentifier,Result_MeasureValueType
19925,Microcystin LR,780.0,ug/L,Final,Actual
11013,Microcystin LR,760.0,ug/L,Final,Actual
76064,Microcystin LR,760.0,ug/L,Final,Actual
56531,Microcystin LR,750.0,ug/L,Final,Actual
362,Microcystin LR,730.0,ug/L,Final,Actual
17648,Microcystin LR,600.0,ug/L,Final,Actual
17671,Microcystin LR,600.0,ug/L,Final,Actual
2367,Microcystin LR,490.0,ug/L,Final,Actual
1049,Microcystin LR,450.0,ug/L,Final,Actual
34048,Microcystin LR,410.0,ug/L,Final,Actual


### Florida Cyanotoxin Outlier Assessment

The IQR method identified statistical outliers across multiple cyanotoxin
characteristics. Because the dataset contains different cyanotoxins with
different concentration distributions, IQR bounds were calculated separately
for each characteristic.

The largest measurements were investigated using their cyanotoxin
characteristic, sampling location, measurement unit, measurement status, and
measurement value type. The extreme values were reported as final actual
measurements in µg/L, with no obvious indication of a data-entry or unit
error. No negative measurements were found.

Therefore, the identified IQR outliers were retained because they were not
determined to be clearly erroneous and may represent legitimate elevated
cyanotoxin concentrations.

habsos_cellcounts.csv

In [38]:
print("Shape:", habsos.shape)

Shape: (219152, 26)


In [39]:
habsos.head()

,OBJECTID,DESCRIPTION,LATITUDE,LONGITUDE,STATE_ID,SAMPLE_DATE,SAMPLE_DEPTH,GENUS,SPECIES,CATEGORY,...,WATER_TEMP,WATER_TEMP_UNIT,WATER_TEMP_QA,WIND_DIR,WIND_DIR_UNIT,WIND_DIR_QA,WIND_SPEED,WIND_SPEED_UNIT,WIND_SPEED_QA,QA_COMMENT
0,1,Tom Adams Bridge (Lemon Bay),26.93450,-82.3535,FL,1953-08-19 00:00:00,0.5,Karenia,brevis,medium,...,NaN,NaN,9,NaN,NaN,9,NaN,NaN,9,Missing/invalid time of day
1,2,Naples Pier,26.13163,-81.8063,FL,1953-08-26 00:00:00,0.5,Karenia,brevis,not observed,...,NaN,NaN,9,NaN,NaN,9,NaN,NaN,9,Missing/invalid time of day
2,3,Piney Point; Lee County,26.53340,-81.9796,FL,1953-08-26 00:00:00,0.5,Karenia,brevis,not observed,...,NaN,NaN,9,NaN,NaN,9,NaN,NaN,9,Missing/invalid time of day
3,4,Cortez Bridge,27.46840,-82.6937,FL,1953-08-26 00:00:00,0.5,Karenia,brevis,medium,...,NaN,NaN,9,NaN,NaN,9,NaN,NaN,9,Missing/invalid time of day
4,5,Sarasota Bay; 3 mi North of bridge,27.37750,-82.5708,FL,1953-09-03 00:00:00,0.5,Karenia,brevis,high,...,NaN,NaN,9,NaN,NaN,9,NaN,NaN,9,Missing/invalid time of day


In [40]:
habsos.columns[habsos.isna().all()]

Index(['WIND_DIR', 'WIND_DIR_UNIT'], dtype='object')

In [41]:
habsos.isna().sum().sort_values(ascending=False)

,0
WIND_DIR,219152
WIND_DIR_UNIT,219152
WIND_SPEED_UNIT,218836
WIND_SPEED,218836
QA_COMMENT,166778
WATER_TEMP,107145
WATER_TEMP_UNIT,107145
SALINITY,105608
SALINITY_UNIT,105608
SAMPLE_DEPTH,6220


In [42]:
habsos.dtypes

,0
OBJECTID,int64
DESCRIPTION,object
LATITUDE,float64
LONGITUDE,float64
STATE_ID,object
SAMPLE_DATE,object
SAMPLE_DEPTH,float64
GENUS,object
SPECIES,object
CATEGORY,object


In [43]:
# Remove completely empty columns.
habsos = habsos.dropna(
    axis=1,
    how="all"
)

# WIND_SPEED and WIND_SPEED_UNIT do not contain usable information.
habsos = habsos.drop(
    columns=["WIND_SPEED", "WIND_SPEED_UNIT"]
)

print("Shape after column cleaning:", habsos.shape)

Shape after column cleaning: (219152, 22)


In [44]:
# Check remaining missing values after column cleaning.
habsos.isna().sum().sort_values(ascending=False)

,0
QA_COMMENT,166778
WATER_TEMP_UNIT,107145
WATER_TEMP,107145
SALINITY,105608
SALINITY_UNIT,105608
SAMPLE_DEPTH,6220
DESCRIPTION,3088
CATEGORY,856
SAMPLE_DATE,0
LATITUDE,0


### HABSOS Missing Values

Water temperature and salinity are sequential environmental measurements.
Short gaps can therefore be estimated from surrounding observations using linear
interpolation.

A limit of two observations is used so that only short gaps are filled.

Categorical variables are filled with `"Unknown"`.

If both salinity and water temperature are missing for an observation, the
observation cannot contribute useful environmental measurements and is removed.

Sample depth is not a sequential measurement in this context, so missing depth
values are represented using the median depth rather than interpolation.

In [45]:
# Calculate the percentage of missing values for the remaining
# environmental measurements.
missing_measurements = habsos[
    ["WATER_TEMP", "SALINITY"]
].isna().sum()

missing_percent = (
    missing_measurements / len(habsos) * 100
)

pd.DataFrame({
    "Missing_Count": missing_measurements,
    "Missing_Percent": missing_percent
})

,Missing_Count,Missing_Percent
WATER_TEMP,107145,48.890724
SALINITY,105608,48.189385


In [46]:
# Check how many observations occur at each geographic sampling location.
location_counts = (
    habsos
    .groupby(["LATITUDE", "LONGITUDE"])
    .size()
    .describe()
)

location_counts

,0
count,37507.000000
mean,5.842963
std,52.230972
min,1.000000
25%,1.000000
50%,1.000000
75%,2.000000
max,5314.000000


In [47]:
# Convert the sampling date to a datetime format.
habsos["SAMPLE_DATE"] = pd.to_datetime(habsos["SAMPLE_DATE"])

# Sort observations by location and sampling date.
habsos = habsos.sort_values(
    ["LATITUDE", "LONGITUDE", "SAMPLE_DATE"]
)

# Count repeated observations at each location.
repeated_locations = (
    habsos
    .groupby(["LATITUDE", "LONGITUDE"])
    .size()
)

# Show how many locations have more than one observation.
repeated_locations[repeated_locations > 1].describe()

,0
count,10682.000000
mean,18.004774
std,96.812812
min,2.000000
25%,2.000000
50%,3.000000
75%,8.000000
max,5314.000000


In [48]:
# Find missing values that are between two valid observations
# at the same sampling location.
internal_gaps = {}

for column in ["WATER_TEMP", "SALINITY"]:
    missing = habsos[column].isna()

    gaps = (
        missing
        & missing.groupby(
            [habsos["LATITUDE"], habsos["LONGITUDE"]]
        ).shift(1).eq(False)
        & missing.groupby(
            [habsos["LATITUDE"], habsos["LONGITUDE"]]
        ).shift(-1).eq(False)
    )

    internal_gaps[column] = gaps.sum()

internal_gaps

{'WATER_TEMP': np.int64(3519), 'SALINITY': np.int64(3969)}

In [49]:
# Check the time gaps around the missing values.
for column in ["WATER_TEMP", "SALINITY"]:

    missing = habsos[column].isna()

    before_date = (
        habsos.groupby(["LATITUDE", "LONGITUDE"])["SAMPLE_DATE"]
        .shift(1)
    )

    after_date = (
        habsos.groupby(["LATITUDE", "LONGITUDE"])["SAMPLE_DATE"]
        .shift(-1)
    )

    before_valid = (
        habsos.groupby(["LATITUDE", "LONGITUDE"])[column]
        .shift(1)
        .notna()
    )

    after_valid = (
        habsos.groupby(["LATITUDE", "LONGITUDE"])[column]
        .shift(-1)
        .notna()
    )

    gaps = missing & before_valid & after_valid

    time_gaps = (
        after_date[gaps] - before_date[gaps]
    ).dt.days

    print(column)
    print(time_gaps.describe())

WATER_TEMP
count     3519.000000
mean       163.944018
std        703.437612
min          0.000000
25%         13.000000
50%         26.000000
75%         64.000000
max      14080.000000
Name: SAMPLE_DATE, dtype: float64
SALINITY
count     3969.000000
mean       151.871000
std        616.645858
min          0.000000
25%         13.000000
50%         22.000000
75%         63.000000
max      10466.000000
Name: SAMPLE_DATE, dtype: float64


In [50]:
# Count missing values that occur at each sampling location.
missing_by_location = (
    habsos
    .groupby(["LATITUDE", "LONGITUDE"])[["WATER_TEMP", "SALINITY"]]
    .apply(lambda x: x.isna().sum())
)

# Show locations where at least one measurement is missing.
missing_by_location[
    (missing_by_location["WATER_TEMP"] > 0) |
    (missing_by_location["SALINITY"] > 0)
].head(20)

WATER_TEMP  SALINITY
LATITUDE LONGITUDE                      
24.1011  -82.95608           1         1
24.1864  -81.61270           1         1
         -81.61170           1         1
24.1994  -81.64830           1         1
24.2010  -82.20600           0         1
24.2290  -81.98600           0         1
24.2500  -81.63834           1         1
24.2510  -81.95220           0         1
         -81.50000           0         1
24.2520  -82.97500           0         1
24.2700  -82.50000           0         1
         -82.12000           0         1
24.2800  -81.83000           0         1
24.3090  -82.00000           0         1
24.3200  -82.28000           0         1
24.3280  -82.82500           0         1
24.3300  -83.32000           0         1
         -83.17000           0         1
         -82.20000           0         1
24.3348  -81.83210           2         2

In [51]:
# Fill missing categorical metadata with "Unknown".
categorical_cols = habsos.select_dtypes(
    include=["object"]
).columns

habsos[categorical_cols] = habsos[
    categorical_cols
].fillna("Unknown")

In [52]:
# Remove observations missing both major environmental measurements.
habsos = habsos.dropna(
    subset=["SALINITY", "WATER_TEMP"],
    how="all"
)

In [53]:
# Look at the distribution of sample depth.
habsos["SAMPLE_DEPTH"].describe()

,SAMPLE_DEPTH
count,119334.000000
mean,1.980429
std,8.742694
min,0.000000
25%,0.500000
50%,0.500000
75%,0.500000
max,1000.300000


In [54]:
print("Remaining missing values:")
print(
    habsos.isna().sum()[
        habsos.isna().sum() > 0
    ]
)

Remaining missing values:
SAMPLE_DEPTH     4279
SALINITY        10069
WATER_TEMP      11606
dtype: int64


### HABSOS Outlier Detection

IQR is applied separately to salinity, water temperature, and cell counts.

Cell counts are strongly right-skewed, so a log transformation is used before
IQR screening. The original `CELLCOUNT` values are retained.

In [55]:
# -----------------------------
# Salinity
# -----------------------------

Q1 = habsos["SALINITY"].quantile(0.25)
Q3 = habsos["SALINITY"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Salinity")
print(f"Q1: {Q1:.3f}")
print(f"Q3: {Q3:.3f}")
print(f"IQR: {IQR:.3f}")
print(f"Lower bound: {lower_bound:.3f}")
print(f"Upper bound: {upper_bound:.3f}")

salinity_outliers = habsos[
    (habsos["SALINITY"] < lower_bound) |
    (habsos["SALINITY"] > upper_bound)
]

print("Number of IQR outliers:", len(salinity_outliers))

display(
    salinity_outliers[
        [
            "OBJECTID",
            "SALINITY",
            "SALINITY_UNIT",
            "SALINITY_QA",
            "QA_COMMENT"
        ]
    ].sort_values(
        "SALINITY",
        ascending=False
    ).head(20)
)

Salinity
Q1: 28.510
Q3: 35.010
IQR: 6.500
Lower bound: 18.760
Upper bound: 44.760
Number of IQR outliers: 6885


,OBJECTID,SALINITY,SALINITY_UNIT,SALINITY_QA,QA_COMMENT
204621,597495,308.00,ppt,4,Unknown
183985,406328,222.70,ppt,4,WDIR unexpected data in input;CATEGORIES from ...
196276,580077,72.00,ppt,4,WTEMP inferred as Fahrenheit based on value an...
182167,346512,57.80,ppt,4,SALINITY value (57.80) exceeds valid range (0....
4921,4922,57.00,ppt,4,SALINITY value (57.00) exceeds valid range (0....
5025,5026,56.00,ppt,4,SALINITY value (56.00) exceeds valid range (0....
4323,4324,55.61,ppt,4,SALINITY value (55.61) exceeds valid range (0....
4736,4737,55.60,ppt,4,SALINITY value (55.60) exceeds valid range (0....
4739,4740,54.40,ppt,4,SALINITY value (54.40) exceeds valid range (0....
177050,182150,54.32,ppt,4,CATEGORIES from provider;Converted from feet t...


In [56]:
# Remove values above the selected physically plausible salinity range.
habsos = habsos.drop(
    habsos[
        habsos["SALINITY"] > 50
    ].index
)

print("Maximum salinity after cleaning:")
print(habsos["SALINITY"].max())

Maximum salinity after cleaning:
50.0


In [57]:
# -----------------------------
# Water temperature
# -----------------------------

Q1 = habsos["WATER_TEMP"].quantile(0.25)
Q3 = habsos["WATER_TEMP"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Water temperature")
print(f"Q1: {Q1:.3f}")
print(f"Q3: {Q3:.3f}")
print(f"IQR: {IQR:.3f}")
print(f"Lower bound: {lower_bound:.3f}")
print(f"Upper bound: {upper_bound:.3f}")

water_temp_outliers = habsos[
    (habsos["WATER_TEMP"] < lower_bound) |
    (habsos["WATER_TEMP"] > upper_bound)
]

print("Number of IQR outliers:", len(water_temp_outliers))

display(
    water_temp_outliers[
        [
            "OBJECTID",
            "WATER_TEMP",
            "WATER_TEMP_UNIT",
            "WATER_TEMP_QA",
            "QA_COMMENT"
        ]
    ].sort_values(
        "WATER_TEMP",
        ascending=False
    ).head(20)
)

Water temperature
Q1: 21.000
Q3: 29.000
IQR: 8.000
Lower bound: 9.000
Upper bound: 41.000
Number of IQR outliers: 829


,OBJECTID,WATER_TEMP,WATER_TEMP_UNIT,WATER_TEMP_QA,QA_COMMENT
184352,406695,1503.888889,deg. C,4,WTEMP value (1503.89) exceeds valid range (-2....
184302,406645,568.333333,deg. C,4,WTEMP value (568.33) exceeds valid range (-2.0...
213213,611134,149.111111,deg. C,4,Unknown
186448,434867,100.000000,deg. C,4,WTEMP value (100.00) exceeds valid range (-2.0...
212474,609991,8.940000,deg. C,1,Unknown
217299,616954,8.900000,deg. C,1,Unknown
151036,152747,8.900000,deg. C,1,Unknown
119588,120373,8.900000,deg. C,1,Unknown
217300,616955,8.900000,deg. C,1,Unknown
217297,616952,8.900000,deg. C,1,Unknown


In [58]:
# Examine the distribution of cell counts.
habsos["CELLCOUNT"].describe()

,CELLCOUNT
count,1.235960e+05
mean,7.841066e+04
std,1.382431e+06
min,0.000000e+00
25%,0.000000e+00
50%,0.000000e+00
75%,0.000000e+00
max,1.894560e+08


In [59]:
# Remove water temperatures outside the selected physically plausible range.
habsos = habsos.drop(
    habsos[
        (habsos["WATER_TEMP"] > 40) |
        (habsos["WATER_TEMP"] < -2)
    ].index
)

print("Water temperature range after cleaning:")
print(habsos["WATER_TEMP"].describe())

Water temperature range after cleaning:
count    111986.000000
mean         24.520341
std           5.376475
min           0.000000
25%          21.000000
50%          25.200000
75%          29.000000
max          33.000000
Name: WATER_TEMP, dtype: float64


In [60]:
# -----------------------------
# Cell count
# -----------------------------

# Log-transform cell counts to reduce strong right-skew.
habsos["LOG_CELLCOUNT"] = np.log1p(
    habsos["CELLCOUNT"]
)

Q1 = habsos["LOG_CELLCOUNT"].quantile(0.25)
Q3 = habsos["LOG_CELLCOUNT"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Log cell count")
print(f"Q1: {Q1:.3f}")
print(f"Q3: {Q3:.3f}")
print(f"IQR: {IQR:.3f}")
print(f"Lower bound: {lower_bound:.3f}")
print(f"Upper bound: {upper_bound:.3f}")

cellcount_outliers = habsos[
    (habsos["LOG_CELLCOUNT"] < lower_bound) |
    (habsos["LOG_CELLCOUNT"] > upper_bound)
]

print("Number of IQR outliers:", len(cellcount_outliers))

display(
    cellcount_outliers[
        [
            "OBJECTID",
            "CELLCOUNT",
            "LOG_CELLCOUNT",
            "CELLCOUNT_QA",
            "QA_COMMENT"
        ]
    ].sort_values(
        "CELLCOUNT",
        ascending=False
    ).head(30)
)

Log cell count
Q1: 0.000
Q3: 0.000
IQR: 0.000
Lower bound: 0.000
Upper bound: 0.000
Number of IQR outliers: 23963


,OBJECTID,CELLCOUNT,LOG_CELLCOUNT,CELLCOUNT_QA,QA_COMMENT
175953,181053,189456000.0,19.059667,1,CATEGORIES from provider;WTEMP inferred as Fah...
175941,181040,138000000.0,18.742764,1,CATEGORIES from provider;WTEMP inferred as Fah...
20859,20906,107000000.0,18.488339,1,Unknown
38100,38201,106563000.0,18.484247,1,Unknown
175952,181052,100000000.0,18.420681,1,CATEGORIES from provider;WTEMP inferred as Fah...
10088,10090,100000000.0,18.420681,1,Unknown
35611,35712,81700000.0,18.218565,1,Missing/invalid time of day
18745,18792,72540000.0,18.099649,1,Unknown
34909,35010,68500000.0,18.042344,1,Unknown
109916,110510,68420000.0,18.041176,1,Unknown


In [61]:
# Examine the distribution of positive CELLCOUNT values.
positive_cellcount = habsos.loc[
    habsos["CELLCOUNT"] > 0, "CELLCOUNT"
]

print("Number of positive CELLCOUNT values:", len(positive_cellcount))
print("\nPositive CELLCOUNT summary:")
print(positive_cellcount.describe())

Number of positive CELLCOUNT values: 23963

Positive CELLCOUNT summary:
count    2.396300e+04
mean     4.044249e+05
std      3.118589e+06
min      1.000000e+00
25%      1.333000e+03
50%      1.550000e+04
75%      1.300000e+05
max      1.894560e+08
Name: CELLCOUNT, dtype: float64


In [62]:
# Apply log1p only to positive CELLCOUNT values for outlier screening.
positive_log_cellcount = np.log1p(positive_cellcount)

# Calculate IQR bounds.
Q1 = positive_log_cellcount.quantile(0.25)
Q3 = positive_log_cellcount.quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Positive CELLCOUNT (log1p transformed)")
print(f"Q1: {Q1:.3f}")
print(f"Q3: {Q3:.3f}")
print(f"IQR: {IQR:.3f}")
print(f"Lower bound: {lower_bound:.3f}")
print(f"Upper bound: {upper_bound:.3f}")

# Identify positive CELLCOUNT values outside the log-IQR bounds.
positive_outliers = positive_cellcount[
    (positive_log_cellcount < lower_bound) |
    (positive_log_cellcount > upper_bound)
]

print("Number of positive IQR outliers:", len(positive_outliers))

# Convert the upper log bound back to the original CELLCOUNT scale.
original_upper_bound = np.expm1(upper_bound)

print(f"Approximate upper bound on original scale: {original_upper_bound:,.0f}")

Positive CELLCOUNT (log1p transformed)
Q1: 7.196
Q3: 11.775
IQR: 4.579
Lower bound: 0.327
Upper bound: 18.644
Number of positive IQR outliers: 2
Approximate upper bound on original scale: 125,064,193


In [63]:
# Inspect the positive CELLCOUNT outliers and their quality information.
cellcount_outliers = habsos[
    habsos["CELLCOUNT"] > original_upper_bound
]

display(
    cellcount_outliers[
        [
            "OBJECTID",
            "CELLCOUNT",
            "CELLCOUNT_UNIT",
            "CELLCOUNT_QA",
            "QA_COMMENT"
        ]
    ].sort_values("CELLCOUNT", ascending=False)
)

,OBJECTID,CELLCOUNT,CELLCOUNT_UNIT,CELLCOUNT_QA,QA_COMMENT
175953,181053,189456000.0,cells/L,1,CATEGORIES from provider;WTEMP inferred as Fah...
175941,181040,138000000.0,cells/L,1,CATEGORIES from provider;WTEMP inferred as Fah...


### HABSOS Outlier Assessment

Outliers were assessed for the main numerical environmental variables using
the IQR method, while considering the distribution, units, and available QA
information.

For SALINITY, the IQR method identified numerous statistical outliers.
Values above 50 ppt were removed because the dataset's QA comments identified
values above this range as exceeding the valid salinity range.

For WATER_TEMP, the IQR method identified statistical outliers. Values below
-2°C or above 40°C were removed because the dataset's QA comments identified
values outside this range as invalid. Remaining values ranged from 0°C to
33°C.

For CELLCOUNT, the data were highly right-skewed with many zero values.
Positive cell counts were therefore assessed using a log1p transformation.
Only two statistical outliers were identified, at 189,456,000 and
138,000,000 cells/L. These were retained because their units and QA
information were valid and the QA comments did not identify the cell counts
as erroneous.

SAMPLE_DEPTH was also examined and contained an extreme value of 1000.3.
This value was identified as a potential outlier but was not removed during
this outlier-cleaning step because additional information is needed to
determine whether it represents a valid sampling depth.

Overall, statistical outliers were not automatically removed. Values were
removed only when the available QA information provided evidence that they
were invalid or clearly erroneous, while potentially legitimate environmental
extremes were retained.

## MoteMarine_BottomTemp.csv

In [64]:
print("Shape:", motemarine.shape)

Shape: (31729, 9)


In [65]:
print("\nColumns:")
print(motemarine.columns.tolist())


Columns:
['time', 'latitude', 'longitude', 'z', 'sea_water_temperature', 'sea_water_temperature_qc_agg', 'month', 'month_sin', 'month_cos']


In [66]:
print("\nData types:")
print(motemarine.dtypes)


Data types:
time                             object
latitude                        float64
longitude                       float64
z                               float64
sea_water_temperature           float64
sea_water_temperature_qc_agg    float64
month                             int64
month_sin                       float64
month_cos                       float64
dtype: object


In [67]:
print("\nSummary statistics:")
display(motemarine.describe())


Summary statistics:


,latitude,longitude,z,sea_water_temperature,sea_water_temperature_qc_agg,month,month_sin,month_cos
count,3.172900e+04,3.172900e+04,31729.0,3.172900e+04,31729.000000,31729.000000,3.172900e+04,3.172900e+04
mean,2.732920e+01,-8.255640e+01,-2.0,7.667743e-16,1.006366,6.614769,-1.984577e-02,7.680694e-02
std,8.462697e-12,2.209823e-11,0.0,1.000016e+00,0.113498,3.585760,7.149142e-01,6.947200e-01
min,2.732920e+01,-8.255640e+01,-2.0,-2.683630e+00,1.000000,1.000000,-1.000000e+00,-1.000000e+00
25%,2.732920e+01,-8.255640e+01,-2.0,-8.271209e-01,1.000000,3.000000,-8.660254e-01,-5.000000e-01
50%,2.732920e+01,-8.255640e+01,-2.0,6.373729e-02,1.000000,7.000000,-2.449294e-16,6.123234e-17
75%,2.732920e+01,-8.255640e+01,-2.0,9.487492e-01,1.000000,10.000000,5.000000e-01,8.660254e-01
max,2.732920e+01,-8.255640e+01,-2.0,1.675910e+00,4.000000,12.000000,1.000000e+00,1.000000e+00


In [68]:
motemarine.head()

,time,latitude,longitude,z,sea_water_temperature,sea_water_temperature_qc_agg,month,month_sin,month_cos
0,2020-08-13 15:00:00+00:00,27.3292,-82.5564,-2.0,1.249129,1.0,8,-0.866025,-0.5
1,2020-08-13 16:00:00+00:00,27.3292,-82.5564,-2.0,1.278159,1.0,8,-0.866025,-0.5
2,2020-08-13 17:00:00+00:00,27.3292,-82.5564,-2.0,1.307189,1.0,8,-0.866025,-0.5
3,2020-08-13 18:00:00+00:00,27.3292,-82.5564,-2.0,1.318680,1.0,8,-0.866025,-0.5
4,2020-08-13 19:00:00+00:00,27.3292,-82.5564,-2.0,1.342065,1.0,8,-0.866025,-0.5


In [69]:
# Remove the metadata/units row at the top of the dataset.
motemarine = motemarine.iloc[1:].copy()

In [70]:
# Convert time to datetime.
motemarine["time"] = pd.to_datetime(motemarine["time"], errors="coerce")

In [71]:
# Convert environmental measurements to numeric.
measurement_cols = [
    "latitude",
    "longitude",
    "z",
    "sea_water_temperature"
]

motemarine[measurement_cols] = motemarine[
    measurement_cols
].apply(
    pd.to_numeric,
    errors="coerce"
)

In [72]:
# Check the resulting data types.
print(motemarine.dtypes)

time                            datetime64[ns, UTC]
latitude                                    float64
longitude                                   float64
z                                           float64
sea_water_temperature                       float64
sea_water_temperature_qc_agg                float64
month                                         int64
month_sin                                   float64
month_cos                                   float64
dtype: object


In [73]:
# Check missing values.
print(
    motemarine.isna().sum()[
        motemarine.isna().sum() > 0
    ]
)

Series([], dtype: int64)


In [74]:
# Review the numerical variables before outlier assessment.
display(
    motemarine[
        [
            "latitude",
            "longitude",
            "z",
            "sea_water_temperature",
            "sea_water_temperature_qc_agg"
        ]
    ].describe()
)

,latitude,longitude,z,sea_water_temperature,sea_water_temperature_qc_agg
count,3.172800e+04,3.172800e+04,31728.0,31728.000000,31728.000000
mean,2.732920e+01,-8.255640e+01,-2.0,-0.000039,1.006367
std,8.459145e-12,2.209823e-11,0.0,1.000007,0.113500
min,2.732920e+01,-8.255640e+01,-2.0,-2.683630,1.000000
25%,2.732920e+01,-8.255640e+01,-2.0,-0.827121,1.000000
50%,2.732920e+01,-8.255640e+01,-2.0,0.063737,1.000000
75%,2.732920e+01,-8.255640e+01,-2.0,0.948749,1.000000
max,2.732920e+01,-8.255640e+01,-2.0,1.675910,4.000000


### Mote Marine Missing Values

The missing QC unit is metadata and cannot be reliably inferred, so it is left
missing.

Environmental measurement values are not median-imputed because there is not
enough justification to replace an unrecorded observation with a typical value.

In [75]:
print("Remaining missing values:")
print(
    motemarine.isna().sum()[
        motemarine.isna().sum() > 0
    ]
)

Remaining missing values:
Series([], dtype: int64)


### Mote Marine Outlier Detection

IQR is used to screen the bottom-water temperature measurement.

Latitude and longitude are not treated as generic outliers because an extreme
coordinate may simply represent a legitimate sampling location.

In [76]:
# Use bottom-water temperature as the environmental measurement for
# outlier screening.
values = motemarine["sea_water_temperature"].dropna()

Q1 = values.quantile(0.25)
Q3 = values.quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"Q1: {Q1:.3f}")
print(f"Q3: {Q3:.3f}")
print(f"IQR: {IQR:.3f}")
print(f"Lower bound: {lower_bound:.3f}")
print(f"Upper bound: {upper_bound:.3f}")

temperature_outliers = motemarine[
    (motemarine["sea_water_temperature"] < lower_bound) |
    (motemarine["sea_water_temperature"] > upper_bound)
]

print("Number of IQR outliers:", len(temperature_outliers))

display(temperature_outliers.head(20))

Q1: -0.827
Q3: 0.949
IQR: 1.776
Lower bound: -3.491
Upper bound: 3.613
Number of IQR outliers: 0


,time,latitude,longitude,z,sea_water_temperature,sea_water_temperature_qc_agg,month,month_sin,month_cos


### MoteMarine Outlier Assessment

The MoteMarine dataset contained no missing values after removing the
metadata/units row and converting the measurement columns to their appropriate
numeric types.

The IQR method was applied to sea-water temperature. The calculated bounds
were 7.416°C and 42.653°C, while the observed temperatures ranged from
11.421°C to 33.046°C. Therefore, no observations were identified as IQR
outliers.

No temperature values were removed during outlier cleaning because all
observations fell within the calculated IQR bounds.

## NutrientsFL_2006_2025.csv

In [77]:
print("Shape:", nutrientsFL.shape)

Shape: (509050, 18)


In [78]:
# Check data types.
print("Data types:")
print(nutrientsFL.dtypes)

Data types:
DataSource              object
StationID               object
Actual_DataSource       object
Actual_StationID        object
Activity_Start_Date     object
Activity_Start_Time     object
Activity_Type           object
RelativeDepth           object
Activity_Depth         float64
Activity_Depth_Unit     object
Characteristic          object
Result_Value           float64
Result_Unit             object
Value_Qualifier         object
Sample_Fraction         object
MDL                    float64
MDL_Unit                object
Result_Comment          object
dtype: object


In [79]:
nutrientsFL.head()

,DataSource,StationID,Actual_DataSource,Actual_StationID,Activity_Start_Date,Activity_Start_Time,Activity_Type,RelativeDepth,Activity_Depth,Activity_Depth_Unit,Characteristic,Result_Value,Result_Unit,Value_Qualifier,Sample_Fraction,MDL,MDL_Unit,Result_Comment
0,SARASOTA_COASTALCREEK_WQ,ALL,SARASOTA_COASTALCREEK_WQ,ALL,7/28/2006 0:00,11:53:00 AM,Sample,Surface,0.5,m,"BOD, Biochemical oxygen demand",2.16,mg/l,I,Unknown,NaN,Unknown,NaN
1,SARASOTA_COASTALCREEK_WQ,ALL,SARASOTA_COASTALCREEK_WQ,ALL,7/28/2006 0:00,11:53:00 AM,Sample,Surface,0.5,m,"Chlorophyll a, corrected for pheophytin",21.40,ug/l,Unknown,Unknown,NaN,Unknown,NaN
2,SARASOTA_COASTALCREEK_WQ,ALL,SARASOTA_COASTALCREEK_WQ,ALL,7/28/2006 0:00,11:53:00 AM,Sample,Surface,0.5,m,Dissolved oxygen (DO),4.25,mg/l,Unknown,Unknown,NaN,Unknown,NaN
3,SARASOTA_COASTALCREEK_WQ,ALL,SARASOTA_COASTALCREEK_WQ,ALL,7/28/2006 0:00,11:53:00 AM,Sample,Surface,0.5,m,Dissolved oxygen saturation,54.90,percent (%),Unknown,Unknown,NaN,Unknown,NaN
4,SARASOTA_COASTALCREEK_WQ,ALL,SARASOTA_COASTALCREEK_WQ,ALL,7/28/2006 0:00,11:53:00 AM,Sample,Surface,0.5,m,Fecal Coliform,330.00,cfu/100ml,Unknown,Unknown,NaN,Unknown,NaN


In [80]:
missing_values = nutrientsFL.isna().sum()

print(
    missing_values[
        missing_values > 0
    ]
)

Activity_Start_Time     82051
Activity_Depth           1911
Activity_Depth_Unit      2106
Result_Unit             54594
MDL                    211237
Result_Comment         426261
dtype: int64


In [81]:
# Examine the measurement units and check for negative measurement values.
print("\nResult_Unit value counts:")
print(nutrientsFL["Result_Unit"].value_counts(dropna=False).head(20))

print("\nNegative Result_Value values:")
print((nutrientsFL["Result_Value"] < 0).sum())


Result_Unit value counts:
Result_Unit
mg/l           219964
percent (%)     57541
umho            56809
NaN             54594
deg F           50351
ug/l            28640
NTU             20811
cfu/100ml       14389
PCU              5951
Name: count, dtype: int64

Negative Result_Value values:
0


In [82]:
# Calculate missing-value percentages.
null_percent = nutrientsFL.isna().mean() * 100

print(
    null_percent[
        null_percent > 0
    ].sort_values(ascending=False)
)

Result_Comment         83.736568
MDL                    41.496317
Activity_Start_Time    16.118456
Result_Unit            10.724683
Activity_Depth_Unit     0.413712
Activity_Depth          0.375405
dtype: float64


### NutrientsFL Missing Values

The NutrientsFL dataset contains different types of environmental measurements
and sampling metadata.

Because these variables have different meanings, missing values are handled
according to the specific field:

- Categorical metadata → `"Unknown"` when appropriate.
- Primary measurement values → left missing unless the observation is unusable.
- Detection limits → left missing because they are measurement metadata.
- Sampling dates/times → left missing because they cannot be reliably inferred.
- Sampling depth → left missing when it was not recorded.
- Measurement units → left missing when the missing value has a meaningful
  relationship to the measurement, such as dimensionless pH.

In [83]:
# Categorical fields where "Unknown" is an appropriate representation
# of an unreported value.
categorical_fill_cols = [
    "Activity_Type",
    "RelativeDepth",
    "Sample_Fraction",
    "MDL_Unit",
    "Value_Qualifier"
]

nutrientsFL[categorical_fill_cols] = nutrientsFL[
    categorical_fill_cols
].fillna("Unknown")

print("Remaining missing values:")
print(
    nutrientsFL.isna().sum()[
        nutrientsFL.isna().sum() > 0
    ]
)

Remaining missing values:
Activity_Start_Time     82051
Activity_Depth           1911
Activity_Depth_Unit      2106
Result_Unit             54594
MDL                    211237
Result_Comment         426261
dtype: int64


### NutrientsFL Outlier Detection

`Result_Value` contains many different environmental measurements. Applying one
IQR calculation to the entire column would be inappropriate because the values
have different units and distributions.

Instead, IQR is calculated separately for each `Characteristic`.

In [84]:
# Get all unique characteristics that have numerical Result_Value data.
characteristics = nutrientsFL[
    "Characteristic"
].dropna().unique()

print("Number of characteristics:", len(characteristics))
print(characteristics)

Number of characteristics: 21
['BOD, Biochemical oxygen demand'
 'Chlorophyll a, corrected for pheophytin' 'Dissolved oxygen (DO)'
 'Dissolved oxygen saturation' 'Fecal Coliform' 'Nitrogen'
 'Nitrogen, ammonia as N' 'Nitrogen, Kjeldahl'
 'Nitrogen, Nitrite (NO2) + Nitrate (NO3) as N' 'pH' 'Phosphorus as P'
 'Phosphorus, phosphate (PO4) as P' 'Specific conductance'
 'Temperature, water' 'Total Suspended Solids (TSS)' 'Turbidity'
 'True Color' 'Chlorophyll a, uncorrected for pheophytin'
 'Temperature, Water' 'Dissolved Oxygen Saturation' 'Specific Conductance']


In [85]:
# Calculate IQR bounds separately for each characteristic.
iqr_by_characteristic = (
    nutrientsFL
    .groupby("Characteristic")["Result_Value"]
    .agg(
        Q1=lambda x: x.quantile(0.25),
        Q3=lambda x: x.quantile(0.75)
    )
)

iqr_by_characteristic["IQR"] = (
    iqr_by_characteristic["Q3"] - iqr_by_characteristic["Q1"]
)

iqr_by_characteristic["Lower_Bound"] = (
    iqr_by_characteristic["Q1"] - 1.5 * iqr_by_characteristic["IQR"]
)

iqr_by_characteristic["Upper_Bound"] = (
    iqr_by_characteristic["Q3"] + 1.5 * iqr_by_characteristic["IQR"]
)

display(iqr_by_characteristic)

,Q1,Q3,IQR,Lower_Bound,Upper_Bound
Characteristic,,,,,
"BOD, Biochemical oxygen demand",0.81000,1.710,0.90000,-0.540000,3.060000
"Chlorophyll a, corrected for pheophytin",2.44000,10.500,8.06000,-9.650000,22.590000
"Chlorophyll a, uncorrected for pheophytin",2.18000,15.300,13.12000,-17.500000,34.980000
Dissolved Oxygen Saturation,88.00000,104.000,16.00000,64.000000,128.000000
Dissolved oxygen (DO),4.77987,7.200,2.42013,1.149675,10.830195
Dissolved oxygen saturation,70.00000,99.000,29.00000,26.500000,142.500000
Fecal Coliform,100.00000,1400.000,1300.00000,-1850.000000,3350.000000
Nitrogen,0.37500,1.150,0.77500,-0.787500,2.312500
"Nitrogen, Kjeldahl",0.39000,1.090,0.70000,-0.660000,2.140000


In [86]:
# Apply IQR outlier detection separately to each characteristic.
#
# Result_Value contains measurements with different units and distributions,
# so one overall IQR calculation would not be appropriate. Instead, each
# characteristic gets its own Q1, Q3, IQR, and outlier boundaries.

for characteristic in characteristics:

    # Get the Result_Value measurements for this characteristic.
    values = nutrientsFL.loc[
        nutrientsFL["Characteristic"] == characteristic,
        "Result_Value"
    ].dropna()

    # Skip characteristics that do not have numerical measurements.
    if len(values) == 0:
        continue

    # Calculate the first and third quartiles.
    Q1 = values.quantile(0.25)
    Q3 = values.quantile(0.75)

    # Calculate the interquartile range.
    IQR = Q3 - Q1

    # Calculate the IQR outlier boundaries.
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Identify measurements outside the IQR boundaries.
    outliers = nutrientsFL.loc[
        (nutrientsFL["Characteristic"] == characteristic) &
        (
            (nutrientsFL["Result_Value"] < lower_bound) |
            (nutrientsFL["Result_Value"] > upper_bound)
        )
    ]

    print("\n" + "=" * 70)
    print(characteristic)
    print("=" * 70)

    print(f"Number of observations: {len(values)}")
    print(f"Q1: {Q1:.3f}")
    print(f"Q3: {Q3:.3f}")
    print(f"IQR: {IQR:.3f}")
    print(f"Lower bound: {lower_bound:.3f}")
    print(f"Upper bound: {upper_bound:.3f}")
    print(f"Number of IQR outliers: {len(outliers)}")

    # Display the largest outliers for inspection.
    if len(outliers) > 0:
        display(
            outliers[
                [
                    "Activity_Start_Date",
                    "StationID",
                    "Result_Value",
                    "Result_Unit",
                    "Value_Qualifier",
                    "Result_Comment"
                ]
            ].sort_values(
                "Result_Value",
                ascending=False
            ).head(20)
        )


BOD, Biochemical oxygen demand
Number of observations: 25640
Q1: 0.810
Q3: 1.710
IQR: 0.900
Lower bound: -0.540
Upper bound: 3.060
Number of IQR outliers: 2255


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
140537,4/6/2009 0:00,MY-A,20.0,mg/l,I,NaN
1113,6/22/2011 0:00,ALL,20.0,mg/l,Unknown,NaN
1114,6/22/2011 0:00,ALL,20.0,mg/l,Unknown,NaN
9639,5/20/2015 0:00,ALL-3,19.6,mg/l,Unknown,NaN
9640,5/20/2015 0:00,ALL-3,19.6,mg/l,Unknown,NaN
96052,3/21/2011 0:00,GOT-D,19.6,mg/l,Unknown,NaN
96053,3/21/2011 0:00,GOT-D,19.6,mg/l,Unknown,NaN
5378,6/22/2011 0:00,ALL-2,17.3,mg/l,Unknown,NaN
5379,6/22/2011 0:00,ALL-2,17.3,mg/l,Unknown,NaN
32175,5/11/2020 14:14,C341-17th,17.3,mg/l,Unknown,NaN



Chlorophyll a, corrected for pheophytin
Number of observations: 26377
Q1: 2.440
Q3: 10.500
IQR: 8.060
Lower bound: -9.650
Upper bound: 22.590
Number of IQR outliers: 2556


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
194266,2/15/2018 0:00,SW-Head,4333.0,ug/l,Unknown,NaN
200029,6/17/2021 9:22,VG-1,2651.0,ug/l,Unknown,NaN
194304,4/17/2018 0:00,SW-Head,2403.0,ug/l,Unknown,NaN
139451,5/9/2022 9:27,ML,1573.0,ug/l,Unknown,NaN
30702,7/23/2022 12:02,BW-Tail,793.0,ug/l,Unknown,NaN
202071,6/17/2021 9:54,VG-4,674.0,ug/l,Unknown,NaN
200015,5/20/2021 9:35,VG-1,667.0,ug/l,Unknown,NaN
200154,5/23/2022 9:10,VG-1,611.0,ug/l,Unknown,NaN
202186,5/23/2022 9:26,VG-4,530.0,ug/l,Unknown,NaN
199881,7/16/2020 9:26,VG-1,518.0,ug/l,Unknown,NaN



Dissolved oxygen (DO)
Number of observations: 27081
Q1: 4.780
Q3: 7.200
IQR: 2.420
Lower bound: 1.150
Upper bound: 10.830
Number of IQR outliers: 1285


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
180968,5/14/2012 0:00,RBW-H,19.85,mg/l,Unknown,NaN
180969,5/14/2012 0:00,RBW-H,19.85,mg/l,Unknown,NaN
180784,10/11/2011 0:00,RBW-H,19.23,mg/l,Unknown,NaN
180783,10/11/2011 0:00,RBW-H,19.23,mg/l,Unknown,NaN
164230,1/20/2020 0:00,NOR,19.20,mg/l,Unknown,NaN
218054,3/21/2011 0:00,WOD,19.03,mg/l,Unknown,NaN
218055,3/21/2011 0:00,WOD,19.03,mg/l,Unknown,NaN
30252,4/18/2019 0:00,BW-Tail,18.60,mg/l,Unknown,NaN
153006,2/21/2013 0:00,MY-E,18.31,mg/l,Unknown,NaN
153005,2/21/2013 0:00,MY-E,18.31,mg/l,Unknown,NaN



Dissolved oxygen saturation
Number of observations: 45782
Q1: 70.000
Q3: 99.000
IQR: 29.000
Lower bound: 26.500
Upper bound: 142.500
Number of IQR outliers: 2735


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
145175,9/10/2012 0:00,MY-B,284.00000,percent (%),Unknown,NaN
145174,9/10/2012 0:00,MY-B,284.00000,percent (%),Unknown,NaN
181236,3/22/2013 0:00,RBW-H,262.40000,percent (%),Unknown,NaN
181235,3/22/2013 0:00,RBW-H,262.40000,percent (%),Unknown,NaN
180971,5/14/2012 0:00,RBW-H,252.50000,percent (%),Unknown,NaN
180970,5/14/2012 0:00,RBW-H,252.50000,percent (%),Unknown,NaN
180786,10/11/2011 0:00,RBW-H,247.50000,percent (%),Unknown,NaN
180785,10/11/2011 0:00,RBW-H,247.50000,percent (%),Unknown,NaN
30253,4/18/2019 0:00,BW-Tail,233.20000,percent (%),Unknown,NaN
160241,3/13/2013 0:00,MYA-C,228.04698,percent (%),Unknown,Water Institute Calculated



Fecal Coliform
Number of observations: 14389
Q1: 100.000
Q3: 1400.000
IQR: 1300.000
Lower bound: -1850.000
Upper bound: 3350.000
Number of IQR outliers: 1901


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
107468,8/29/2012 0:00,HUD-3,194000.0,cfu/100ml,B,NaN
107467,8/29/2012 0:00,HUD-3,194000.0,cfu/100ml,B,NaN
103018,6/29/2011 0:00,HUD-2,191000.0,cfu/100ml,B,NaN
103019,6/29/2011 0:00,HUD-2,191000.0,cfu/100ml,B,NaN
172101,5/12/2025 10:49,PH-SG,190000.0,cfu/100ml,B,25050535
35379,6/12/2023 11:36,C498-Lin,176000.0,cfu/100ml,B,NaN
32988,7/21/2025 14:33,C341-17th,174000.0,cfu/100ml,B,25071120
126470,11/16/2006 0:00,MAT-2,170000.0,cfu/100ml,Unknown,NaN
33002,8/11/2025 14:38,C341-17th,167000.0,cfu/100ml,B,NaN
138251,8/11/2025 12:59,MB-Ger,160000.0,cfu/100ml,B,NaN



Nitrogen
Number of observations: 24838
Q1: 0.375
Q3: 1.150
IQR: 0.775
Lower bound: -0.787
Upper bound: 2.312
Number of IQR outliers: 915


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
28880,7/23/2022 11:46,BW-Mid,54.606,mg/l,Unknown,Water Institute Calculated: TKN + NOx
26769,7/23/2022 11:26,BW-Head,44.407,mg/l,Unknown,Water Institute Calculated: TKN + NOx
193640,9/8/2014 0:00,SW-Head,23.807,mg/l,Unknown,NaN
26905,7/24/2023 12:57,BW-Head,21.161,mg/l,Unknown,Water Institute Calculated: TKN + NOx
31757,10/9/2017 9:57,C341-17th,20.653,mg/l,Unknown,Water Institute Calculated: TKN + NOx
196286,7/23/2022 12:35,SW-Mid,20.506,mg/l,Unknown,Water Institute Calculated: TKN + NOx
32010,10/8/2018 0:00,C341-17th,20.278,mg/l,Unknown,Water Institute Calculated: TKN + NOx
31523,11/14/2016 12:50,C341-17th,19.212,mg/l,Unknown,Water Institute Calculated: TKN + NOx
31506,11/14/2016 0:00,C341-17th,19.212,mg/l,Unknown,Water Institute Calculated: TKN + NOx
31779,11/13/2017 10:05,C341-17th,18.938,mg/l,Unknown,Water Institute Calculated: TKN + NOx



Nitrogen, ammonia as N
Number of observations: 14732
Q1: 0.015
Q3: 0.125
IQR: 0.110
Lower bound: -0.150
Upper bound: 0.290
Number of IQR outliers: 1300


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
31507,11/14/2016 0:00,C341-17th,16.80,mg/l,Unknown,NaN
31478,10/17/2016 0:00,C341-17th,16.50,mg/l,Unknown,NaN
31768,11/13/2017 0:00,C341-17th,13.70,mg/l,Unknown,NaN
32263,11/10/2020 14:37,C341-17th,13.60,mg/l,Unknown,NaN
32277,12/7/2020 14:46,C341-17th,13.10,mg/l,Unknown,NaN
31818,1/15/2018 0:00,C341-17th,12.00,mg/l,Unknown,NaN
31709,8/16/2017 0:00,C341-17th,11.10,mg/l,Unknown,NaN
32931,3/10/2025 13:52,C341-17th,9.77,mg/l,Unknown,NaN
32027,11/12/2018 0:00,C341-17th,8.78,mg/l,Unknown,NaN
32235,9/8/2020 14:25,C341-17th,8.46,mg/l,Unknown,NaN



Nitrogen, Kjeldahl
Number of observations: 27473
Q1: 0.390
Q3: 1.090
IQR: 0.700
Lower bound: -0.660
Upper bound: 2.140
Number of IQR outliers: 1084


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
31524,11/14/2016 12:50,C341-17th,18.9,mg/l,,NaN
31781,11/13/2017 10:05,C341-17th,18.9,mg/l,Unknown,NaN
31508,11/14/2016 0:00,C341-17th,18.9,mg/l,Unknown,NaN
31780,11/13/2017 10:05,C341-17th,18.9,mg/l,Unknown,NaN
31525,11/14/2016 12:50,C341-17th,18.9,mg/l,,NaN
31493,10/17/2016 12:19,C341-17th,18.4,mg/l,Unknown,NaN
31492,10/17/2016 12:19,C341-17th,18.4,mg/l,Unknown,NaN
32264,11/10/2020 14:37,C341-17th,17.9,mg/l,Unknown,NaN
200183,7/23/2022 9:15,VG-1,16.0,mg/l,Unknown,NaN
32278,12/7/2020 14:46,C341-17th,14.5,mg/l,Unknown,NaN



Nitrogen, Nitrite (NO2) + Nitrate (NO3) as N
Number of observations: 26732
Q1: 0.005
Q3: 0.075
IQR: 0.070
Lower bound: -0.100
Upper bound: 0.180
Number of IQR outliers: 3472


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
29622,8/18/2016 13:07,BW-Tail,4.59,mg/l,Unknown,NaN
29623,8/18/2016 13:07,BW-Tail,4.59,mg/l,Unknown,NaN
32413,9/13/2021 14:45,C341-17th,3.96,mg/l,Unknown,NaN
104813,1/22/2018 0:00,HUD-2,3.72,mg/l,Unknown,NaN
32423,10/11/2021 14:28,C341-17th,3.61,mg/l,Unknown,NaN
32135,1/13/2020 13:41,C341-17th,3.12,mg/l,Unknown,NaN
59893,2/21/2018 0:00,CLO,2.90,mg/l,Unknown,NaN
183344,2/17/2016 0:00,RBW-SCG,2.53,mg/l,Unknown,NaN
163624,2/21/2017 0:00,NOR,2.38,mg/l,Unknown,NaN
185170,2/7/2007 0:00,RBW-W,1.88,mg/l,Unknown,NaN



pH
Number of observations: 54594
Q1: 7.600
Q3: 8.010
IQR: 0.410
Lower bound: 6.985
Upper bound: 8.625
Number of IQR outliers: 2790


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
139572,2/13/2023 0:00,ML,10.75,NaN,Unknown,NaN
183628,3/13/2017 0:00,RBW-SCG,10.70,NaN,Unknown,NaN
139123,4/13/2020 0:00,ML,10.05,NaN,Unknown,NaN
198591,12/14/2017 9:37,VEN_GAR-5,9.98,NaN,,NaN
139109,3/9/2020 0:00,ML,9.98,NaN,Unknown,NaN
182853,3/13/2018 9:14,RBW-MIR,9.97,NaN,,NaN
201832,7/18/2019 0:00,VG-4,9.94,NaN,Unknown,NaN
201263,7/21/2016 9:47,VG-4,9.92,NaN,,NaN
182911,8/13/2018 11:03,RBW-MIR,9.90,NaN,,NaN
139033,4/16/2019 0:00,ML,9.90,NaN,Unknown,NaN



Phosphorus as P
Number of observations: 27367
Q1: 0.090
Q3: 0.316
IQR: 0.226
Lower bound: -0.249
Upper bound: 0.655
Number of IQR outliers: 1899


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
193647,9/8/2014 0:00,SW-Head,4.78,mg/l,Unknown,NaN
40556,6/25/2010 0:00,CAT-2,4.61,mg/l,Unknown,NaN
40555,6/25/2010 0:00,CAT-2,4.61,mg/l,Unknown,NaN
111531,5/21/2012 0:00,HWC,4.46,mg/l,Unknown,NaN
111530,5/21/2012 0:00,HWC,4.46,mg/l,Unknown,NaN
35456,3/11/2024 11:38,C498-Lin,4.25,mg/l,Unknown,NaN
118801,3/13/2023 12:02,LAB-Web,4.09,mg/l,Unknown,NaN
102913,2/22/2011 0:00,HUD-2,4.05,mg/l,Unknown,NaN
102914,2/22/2011 0:00,HUD-2,4.05,mg/l,Unknown,NaN
194820,3/21/2022 12:21,SW-Head,3.98,mg/l,Unknown,NaN



Phosphorus, phosphate (PO4) as P
Number of observations: 27309
Q1: 0.020
Q3: 0.204
IQR: 0.184
Lower bound: -0.256
Upper bound: 0.480
Number of IQR outliers: 1881


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
111532,5/21/2012 0:00,HWC,3.72,mg/l,Unknown,NaN
111533,5/21/2012 0:00,HWC,3.72,mg/l,Unknown,NaN
40558,6/25/2010 0:00,CAT-2,3.62,mg/l,Unknown,NaN
40557,6/25/2010 0:00,CAT-2,3.62,mg/l,Unknown,NaN
110432,6/30/2008 0:00,HWC,3.22,mg/l,Unknown,NaN
110902,6/21/2010 0:00,HWC,2.90,mg/l,Unknown,NaN
110903,6/21/2010 0:00,HWC,2.90,mg/l,Unknown,NaN
33342,6/12/2008 0:00,C498-Lin,2.89,mg/l,Unknown,NaN
35457,3/11/2024 11:38,C498-Lin,2.81,mg/l,Unknown,NaN
111204,6/28/2011 0:00,HWC,2.81,mg/l,Unknown,NaN



Specific conductance
Number of observations: 48171
Q1: 1548.195
Q3: 52080.000
IQR: 50531.805
Lower bound: -74249.512
Upper bound: 127877.708
Number of IQR outliers: 0

Temperature, water
Number of observations: 38674
Q1: 69.752
Q3: 84.380
IQR: 14.628
Lower bound: 47.811
Upper bound: 106.322
Number of IQR outliers: 50


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
95693,1/11/2010 0:00,GOT-D,47.768,deg F,Unknown,NaN
134787,1/22/2009 0:00,MB-Ger,47.696,deg F,Unknown,NaN
463252,1/12/2010 0:00,LB-3,47.660,deg F,Unknown,Water Institute Calculated
478471,1/12/2010 0:00,LB-5,47.660,deg F,Unknown,Water Institute Calculated
494450,1/12/2010 0:00,US-3,47.660,deg F,Unknown,Water Institute Calculated
373794,1/12/2010 0:00,14-5,47.660,deg F,Unknown,Water Institute Calculated
131106,1/22/2009 0:00,MB-Fruit,47.624,deg F,Unknown,NaN
373795,1/12/2010 0:00,14-5,47.480,deg F,Unknown,Water Institute Calculated
373796,1/12/2010 0:00,14-5,47.480,deg F,Unknown,Water Institute Calculated
494451,1/12/2010 0:00,US-3,47.480,deg F,Unknown,Water Institute Calculated



Total Suspended Solids (TSS)
Number of observations: 18792
Q1: 2.900
Q3: 10.000
IQR: 7.100
Lower bound: -7.750
Upper bound: 20.650
Number of IQR outliers: 1482


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
32921,2/10/2025 13:51,C341-17th,352.0,mg/l,Unknown,NaN
194561,3/19/2020 14:40,SW-Head,347.0,mg/l,Unknown,NaN
32212,7/13/2020 14:17,C341-17th,282.0,mg/l,Unknown,NaN
193631,7/22/2014 0:00,SW-Head,281.0,mg/l,Unknown,NaN
194578,5/21/2020 13:57,SW-Head,275.0,mg/l,Unknown,NaN
31727,8/16/2017 12:46,C341-17th,270.0,mg/l,Unknown,NaN
31726,8/16/2017 12:46,C341-17th,270.0,mg/l,Unknown,NaN
139500,8/15/2022 9:22,ML,264.0,mg/l,Unknown,NaN
138919,5/15/2018 0:00,ML,240.0,mg/l,Unknown,NaN
182879,5/15/2018 9:28,RBW-MIR,240.0,mg/l,,NaN



Turbidity
Number of observations: 20811
Q1: 2.100
Q3: 4.600
IQR: 2.500
Lower bound: -1.650
Upper bound: 8.350
Number of IQR outliers: 1148


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
223058,6/20/2012 0:00,WOD-2,140.0,NTU,Unknown,NaN
223057,6/20/2012 0:00,WOD-2,140.0,NTU,Unknown,NaN
57349,2/26/2009 0:00,CLO,117.0,NTU,Unknown,NaN
201409,3/16/2017 0:00,VG-4,100.0,NTU,Unknown,NaN
58921,6/26/2014 0:00,CLO,90.0,NTU,Unknown,NaN
193632,7/22/2014 0:00,SW-Head,85.0,NTU,Unknown,NaN
223719,10/15/2014 0:00,WOD-2,80.0,NTU,Unknown,NaN
173529,2/12/2010 0:00,RBS-Wilk,76.0,NTU,Unknown,NaN
57333,1/28/2009 0:00,CLO,76.0,NTU,Unknown,NaN
96423,6/20/2012 0:00,GOT-D,75.0,NTU,Unknown,NaN



True Color
Number of observations: 5951
Q1: 6.000
Q3: 90.000
IQR: 84.000
Lower bound: -120.000
Upper bound: 216.000
Number of IQR outliers: 166


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
194739,7/15/2021 12:19,SW-Head,1000.0,PCU,Unknown,NaN
139431,3/7/2022 9:21,ML,600.0,PCU,Unknown,NaN
139417,2/7/2022 9:09,ML,600.0,PCU,Unknown,NaN
73039,10/24/2022 13:42,CPS-4,450.0,PCU,Unknown,NaN
47822,9/20/2022 14:34,CC-2,450.0,PCU,Unknown,NaN
90005,9/20/2022 9:36,GOT-2,450.0,PCU,Unknown,NaN
79378,11/8/2022 9:51,DPS,450.0,PCU,Unknown,NaN
154870,9/14/2021 11:24,MY-E,400.0,PCU,Unknown,NaN
71627,10/24/2022 13:18,CPS-3,400.0,PCU,Unknown,NaN
151035,10/27/2022 11:08,MY-C,400.0,PCU,Unknown,NaN



Chlorophyll a, uncorrected for pheophytin
Number of observations: 2263
Q1: 2.180
Q3: 15.300
IQR: 13.120
Lower bound: -17.500
Upper bound: 34.980
Number of IQR outliers: 277


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
194276,2/15/2018 12:39,SW-Head,4333.0,ug/l,,NaN
194313,4/17/2018 13:06,SW-Head,2403.0,ug/l,,NaN
2997,2/22/2018 12:01,ALL,517.0,ug/l,,NaN
182870,5/15/2018 9:28,RBW-MIR,479.0,ug/l,,NaN
138616,12/12/2016 10:03,ML,332.0,ug/l,,NaN
199182,7/21/2016 9:30,VG-1,285.0,ug/l,,NaN
198536,10/18/2018 9:15,VEN_GAR-4,281.0,ug/l,,NaN
138445,4/18/2016 9:26,ML,269.0,ug/l,,NaN
199665,8/15/2018 9:00,VG-1,266.0,ug/l,,NaN
26224,7/12/2018 11:17,BW-Head,258.0,ug/l,,NaN



Temperature, Water
Number of observations: 11677
Q1: 70.700
Q3: 85.820
IQR: 15.120
Lower bound: 48.020
Upper bound: 108.500
Number of IQR outliers: 0

Dissolved Oxygen Saturation
Number of observations: 11759
Q1: 88.000
Q3: 104.000
IQR: 16.000
Lower bound: 64.000
Upper bound: 128.000
Number of IQR outliers: 665


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
308596,1/6/2021 14:35,13-1,208.0,percent (%),Unknown,NaN
232190,8/8/2018 13:03,10-1,208.0,percent (%),Unknown,NaN
232180,8/8/2018 13:02,10-1,207.0,percent (%),Unknown,NaN
480950,7/10/2019 12:43,LB-5,202.0,percent (%),Unknown,NaN
316103,8/12/2020 14:07,13-2,184.0,percent (%),Unknown,NaN
507122,8/8/2018 12:46,US-5,179.0,percent (%),Unknown,NaN
301325,7/6/2022 13:57,11-5,179.0,percent (%),Unknown,NaN
507132,8/8/2018 12:47,US-5,178.0,percent (%),Unknown,NaN
256688,5/7/2025 14:03,10-4,174.0,percent (%),Unknown,NaN
256699,5/7/2025 14:04,10-4,172.0,percent (%),Unknown,NaN



Specific Conductance
Number of observations: 8638
Q1: 48300.000
Q3: 53500.000
IQR: 5200.000
Lower bound: 40500.000
Upper bound: 61300.000
Number of IQR outliers: 499


,Activity_Start_Date,StationID,Result_Value,Result_Unit,Value_Qualifier,Result_Comment
316387,9/1/2021 13:59,13-2,40400.0,umho,Unknown,NaN
316381,9/1/2021 13:58,13-2,40400.0,umho,Unknown,NaN
481513,9/1/2021 10:01,LB-5,40400.0,umho,Unknown,NaN
301874,8/7/2024 14:07,11-5,40400.0,umho,Unknown,NaN
481507,9/1/2021 10:00,LB-5,40400.0,umho,Unknown,NaN
438042,10/12/2022 11:38,DR-4,40400.0,umho,Unknown,NaN
431664,9/9/2020 12:14,DR-3,40400.0,umho,Unknown,NaN
399047,9/9/2020 10:26,16-3,40400.0,umho,Unknown,NaN
361679,8/11/2021 12:20,14-3,40400.0,umho,Unknown,NaN
481517,9/1/2021 10:02,LB-5,40400.0,umho,Unknown,NaN


In [87]:
# Remove clearly erroneous pH measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "pH") &
        (nutrientsFL["Result_Value"] >= 22)
    ].index
)

# Remove clearly implausible water-temperature measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Temperature, water") &
        (nutrientsFL["Result_Value"] > 110)
    ].index
)

# Remove clearly implausible dissolved oxygen measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Dissolved oxygen (DO)") &
        (nutrientsFL["Result_Value"] > 20)
    ].index
)

# Remove clearly implausible Kjeldahl nitrogen measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Nitrogen, Kjeldahl") &
        (nutrientsFL["Result_Value"] > 20)
    ].index
)

# Remove clearly implausible nitrite + nitrate measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Nitrogen, Nitrite (NO2) + Nitrate (NO3) as N") &
        (nutrientsFL["Result_Value"] > 5)
    ].index
)

# Remove clearly implausible phosphorus measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Phosphorus as P") &
        (nutrientsFL["Result_Value"] > 5)
    ].index
)

# Remove clearly implausible phosphate measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Phosphorus, phosphate (PO4) as P") &
        (nutrientsFL["Result_Value"] > 5)
    ].index
)

# Remove clearly implausible total suspended solids measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Total Suspended Solids (TSS)") &
        (nutrientsFL["Result_Value"] > 2000)
    ].index
)

# Remove clearly implausible turbidity measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "Turbidity") &
        (nutrientsFL["Result_Value"] > 150)
    ].index
)

# Remove clearly implausible true-color measurements.
nutrientsFL = nutrientsFL.drop(
    nutrientsFL[
        (nutrientsFL["Characteristic"] == "True Color") &
        (nutrientsFL["Result_Value"] > 1000)
    ].index
)

Sarasota_Wind&Temp.csv

In [88]:
sarasota.columns

Index(['time', 'z', 'air_pressure_at_mean_sea_level',
       'air_pressure_at_mean_sea_level_qc_agg', 'dew_point_temperature',
       'dew_point_temperature_qc_agg', 'air_temperature',
       'air_temperature_qc_agg', 'visibility_in_air',
       'visibility_in_air_qc_agg', 'wind_speed', 'wind_speed_qc_agg',
       'wind_from_direction_qc_agg', 'wind_dir_sin', 'wind_dir_cos', 'month',
       'month_sin', 'month_cos'],
      dtype='object')

In [89]:
# Remove the units row
sarasota = sarasota.iloc[1:].reset_index(drop=True)

In [90]:
sarasota.head()

,time,z,air_pressure_at_mean_sea_level,air_pressure_at_mean_sea_level_qc_agg,dew_point_temperature,dew_point_temperature_qc_agg,air_temperature,air_temperature_qc_agg,visibility_in_air,visibility_in_air_qc_agg,wind_speed,wind_speed_qc_agg,wind_from_direction_qc_agg,wind_dir_sin,wind_dir_cos,month,month_sin,month_cos
0,2022-07-11 22:53:00+00:00,0.0,-0.835919,2.0,0.718507,2.0,1.155677,2.0,0.042363,2.0,0.843352,2.0,2.0,-0.642788,-0.766044,7,-0.5,-0.866025
1,2022-07-11 23:53:00+00:00,0.0,-0.766508,2.0,0.908343,2.0,1.046460,2.0,0.042363,2.0,0.617244,2.0,2.0,-0.766044,-0.642788,7,-0.5,-0.866025
2,2022-07-12 00:53:00+00:00,0.0,-0.697097,2.0,0.804796,2.0,0.955446,2.0,0.042363,2.0,0.391136,2.0,2.0,-0.866025,-0.500000,7,-0.5,-0.866025
3,2022-07-12 01:53:00+00:00,0.0,-0.488865,2.0,0.908343,2.0,0.955446,2.0,0.042363,2.0,0.165028,2.0,2.0,-0.984808,-0.173648,7,-0.5,-0.866025
4,2022-07-12 02:53:00+00:00,0.0,-0.350044,2.0,0.908343,2.0,0.955446,2.0,0.042363,2.0,-0.513295,2.0,2.0,-0.939693,-0.342020,7,-0.5,-0.866025


In [91]:
sarasota.shape

(41284, 18)

In [92]:
# Convert the time column from text
sarasota["time"] = pd.to_datetime(sarasota["time"])

In [93]:
sarasota.dtypes

,0
time,"datetime64[ns, UTC]"
z,float64
air_pressure_at_mean_sea_level,float64
air_pressure_at_mean_sea_level_qc_agg,float64
dew_point_temperature,float64
dew_point_temperature_qc_agg,float64
air_temperature,float64
air_temperature_qc_agg,float64
visibility_in_air,float64
visibility_in_air_qc_agg,float64


In [94]:
numeric_columns = [
    "air_pressure_at_mean_sea_level",
    "dew_point_temperature",
    "air_temperature",
    "visibility_in_air",
    "wind_speed",
    "wind_from_direction",
    "wind_speed_of_gust"
]

# So we convert them
for column in numeric_columns:
    sarasota[column] = pd.to_numeric(sarasota[column], errors="coerce")

KeyError: 'wind_from_direction'

In [95]:
# Looking for missing values
sarasota.isnull().sum()

,0
time,0
z,0
air_pressure_at_mean_sea_level,0
air_pressure_at_mean_sea_level_qc_agg,0
dew_point_temperature,0
dew_point_temperature_qc_agg,0
air_temperature,0
air_temperature_qc_agg,0
visibility_in_air,0
visibility_in_air_qc_agg,0


In [96]:
missing_percent = sarasota.isnull().mean() * 100
missing_percent.sort_values(ascending=False)

,0
time,0.0
z,0.0
air_pressure_at_mean_sea_level,0.0
air_pressure_at_mean_sea_level_qc_agg,0.0
dew_point_temperature,0.0
dew_point_temperature_qc_agg,0.0
air_temperature,0.0
air_temperature_qc_agg,0.0
visibility_in_air,0.0
visibility_in_air_qc_agg,0.0


In [97]:
sarasota["time"].diff().value_counts().head(10)

,count
time,
0 days 01:00:00,29609
0 days 00:07:00,907
0 days 00:08:00,512
0 days 00:09:00,487
0 days 00:10:00,456
0 days 00:15:00,407
0 days 00:11:00,375
0 days 00:12:00,358
0 days 00:13:00,319


wind_speed_of_gust has 35,924 missing values, which is more than 75% of the data.

In [98]:
# Drop wind_speed_of_gust
sarasota = sarasota.drop(columns=["wind_speed_of_gust"])

KeyError: "['wind_speed_of_gust'] not found in axis"

In [114]:
# Look for duplicates
sarasota.duplicated().sum()

np.int64(0)

In [115]:
# Sort the data by time
sarasota = sarasota.sort_values("time")
sarasota = sarasota.reset_index(drop=True)

In [116]:
# Missing values

# Interpolation
sarasota["air_temperature"] = sarasota["air_temperature"].interpolate()
sarasota["dew_point_temperature"] = sarasota["dew_point_temperature"].interpolate()
sarasota["wind_speed"] = sarasota["wind_speed"].interpolate()

Air pressure has 7,611 missing values.

In [117]:
sarasota[sarasota["air_pressure_at_mean_sea_level"].isnull()]["time"].describe()

,time
count,0
mean,NaT
min,NaT
25%,NaT
50%,NaT
75%,NaT
max,NaT


In [118]:
sarasota.groupby(sarasota["time"].dt.year)["air_pressure_at_mean_sea_level"].apply(
    lambda x: x.isnull().mean() * 100
)

,air_pressure_at_mean_sea_level
time,
2022,0.0
2023,0.0
2024,0.0
2025,0.0
2026,0.0


In [119]:
sarasota[sarasota["air_pressure_at_mean_sea_level"].isnull()][
    ["time", "air_pressure_at_mean_sea_level"]
].head(20)

,time,air_pressure_at_mean_sea_level


In [120]:
# Filling in the missing pressure values
sarasota["air_pressure_at_mean_sea_level"] = (
    sarasota["air_pressure_at_mean_sea_level"].interpolate()
)

sarasota["air_pressure_at_mean_sea_level"].isnull().sum()


np.int64(0)

In [121]:
# Using interpolation again to fill in the missing values
sarasota["visibility_in_air"] = (
    sarasota["visibility_in_air"].interpolate()
)
sarasota["visibility_in_air"].isnull().sum()

np.int64(0)

In [122]:
sarasota.describe()

,z,air_pressure_at_mean_sea_level,air_pressure_at_mean_sea_level_qc_agg,dew_point_temperature,dew_point_temperature_qc_agg,air_temperature,air_temperature_qc_agg,visibility_in_air,visibility_in_air_qc_agg,wind_speed,wind_speed_qc_agg,wind_from_direction_qc_agg,wind_dir_sin,wind_dir_cos,month,month_sin,month_cos
count,41284.0,41284.000000,41284.0,41284.000000,41284.0,41284.000000,41284.0,41284.000000,41284.0,41284.000000,41284.0,41284.0,41284.000000,41284.000000,41284.000000,4.128400e+04,4.128400e+04
mean,0.0,0.000022,2.0,-0.000022,2.0,-0.000028,2.0,-0.000001,2.0,-0.000015,2.0,2.0,0.143446,0.143945,6.570633,-1.592590e-02,1.408011e-02
std,0.0,1.000014,0.0,1.000014,0.0,1.000008,0.0,1.000024,0.0,1.000020,0.0,0.0,0.718496,0.665199,3.481397,7.042281e-01,7.096725e-01
min,0.0,-12.705153,2.0,-5.908512,2.0,-4.305163,2.0,-1.553037,2.0,-1.643834,2.0,2.0,-1.000000,-1.000000,1.000000,-1.000000e+00,-1.000000e+00
25%,0.0,-0.604550,2.0,-0.437770,2.0,-0.555386,2.0,0.042363,2.0,-0.739403,2.0,2.0,-0.500000,-0.500000,4.000000,-8.660254e-01,-5.000000e-01
50%,0.0,0.020147,2.0,0.235287,2.0,0.136320,2.0,0.042363,2.0,-0.061079,2.0,2.0,0.173648,0.173648,7.000000,-2.449294e-16,6.123234e-17
75%,0.0,0.621706,2.0,0.753023,2.0,0.755215,2.0,0.042363,2.0,0.617244,2.0,2.0,0.866025,0.766044,10.000000,5.000000e-01,8.660254e-01
max,0.0,3.536957,2.0,1.581400,2.0,2.975957,2.0,26.713169,2.0,11.922632,2.0,2.0,1.000000,1.000000,12.000000,1.000000e+00,1.000000e+00


In [123]:
# Outlier temperature values
sarasota["air_temperature"].sort_values().head(20)
sarasota["air_temperature"].sort_values(ascending=False).head(20)

,air_temperature
25113,2.975957
19665,2.466278
10281,2.466278
10283,2.375264
10282,2.375264
11348,2.266048
10280,2.266048
11349,2.266048
19661,2.247845
11693,2.175034


-60.6 seems to be an outlier temperature value.

In [124]:
sarasota[sarasota["air_temperature"] < 0][
    ["time", "air_temperature", "dew_point_temperature"]
].head(20)

,time,air_temperature,dew_point_temperature
91,2022-07-14 22:18:00+00:00,-0.063911,0.614960
92,2022-07-14 22:28:00+00:00,-0.063911,0.528671
93,2022-07-14 22:53:00+00:00,-0.063911,0.528671
94,2022-07-14 22:58:00+00:00,-0.063911,0.528671
97,2022-07-15 00:15:00+00:00,-0.063911,0.528671
98,2022-07-15 00:30:00+00:00,-0.063911,0.528671
99,2022-07-15 00:53:00+00:00,-0.063911,0.528671
100,2022-07-15 01:53:00+00:00,-0.154925,0.528671
101,2022-07-15 02:11:00+00:00,-0.063911,0.614960
102,2022-07-15 02:53:00+00:00,-0.063911,0.614960


In [125]:
sarasota.loc[sarasota["air_temperature"] == -60.6]

,time,z,air_pressure_at_mean_sea_level,air_pressure_at_mean_sea_level_qc_agg,dew_point_temperature,dew_point_temperature_qc_agg,air_temperature,air_temperature_qc_agg,visibility_in_air,visibility_in_air_qc_agg,wind_speed,wind_speed_qc_agg,wind_from_direction_qc_agg,wind_dir_sin,wind_dir_cos,month,month_sin,month_cos


In [126]:
sarasota.loc[sarasota["air_temperature"] == 50.0]

,time,z,air_pressure_at_mean_sea_level,air_pressure_at_mean_sea_level_qc_agg,dew_point_temperature,dew_point_temperature_qc_agg,air_temperature,air_temperature_qc_agg,visibility_in_air,visibility_in_air_qc_agg,wind_speed,wind_speed_qc_agg,wind_from_direction_qc_agg,wind_dir_sin,wind_dir_cos,month,month_sin,month_cos


In [127]:
# Investigating the potential outliers

sarasota.loc[9170:9178, ["time", "air_temperature", "dew_point_temperature"]]

,time,air_temperature,dew_point_temperature
9170,2023-05-30 13:53:00+00:00,0.354754,-0.334222
9171,2023-05-30 14:53:00+00:00,0.755215,-0.334222
9172,2023-05-30 15:53:00+00:00,0.955446,-0.437770
9173,2023-05-30 16:53:00+00:00,1.155677,-2.180814
9174,2023-05-30 17:53:00+00:00,1.355908,-0.731153
9175,2023-05-30 18:53:00+00:00,1.355908,-0.627606
9176,2023-05-30 19:53:00+00:00,1.355908,-0.247933
9177,2023-05-30 20:53:00+00:00,1.465125,-0.058097
9178,2023-05-30 21:53:00+00:00,1.264894,-0.247933


In [128]:
# Replacing both outliers
sarasota.loc[sarasota["air_temperature"] == -60.6, "air_temperature"] = np.nan
sarasota.loc[sarasota["air_temperature"] == 50.0, "air_temperature"] = np.nan
sarasota["air_temperature"] = sarasota["air_temperature"].interpolate()
sarasota["air_temperature"].isnull().sum()

np.int64(0)

In [129]:
# Make sure wind direction is reasonable

sarasota["wind_from_direction"].value_counts().sort_index().head(20)
sarasota["wind_from_direction"].value_counts().sort_index().tail(20)

KeyError: 'wind_from_direction'

In [130]:
# More weird temperature values
sarasota[sarasota["air_temperature"] < 0][
    ["time", "air_temperature", "dew_point_temperature"]
].head(20)
sarasota.loc[sarasota["air_temperature"] < 0, "air_temperature"] = np.nan
sarasota["air_temperature"] = sarasota["air_temperature"].interpolate()
sarasota["air_temperature"].isnull().sum()

np.int64(0)

In [131]:
# Investigating the high visibility maximum

sarasota[sarasota["visibility_in_air"] > 100000][
    ["time", "visibility_in_air"]
]
sarasota[sarasota["visibility_in_air"] > 100000]["visibility_in_air"].value_counts()

,count
visibility_in_air,


In [132]:
sarasota.loc[345:350, ["time", "wind_from_direction"]]

# Using forward fill to fill in the missing values
sarasota["wind_from_direction"] = (
    sarasota["wind_from_direction"].ffill()
)

sarasota["wind_from_direction"].isnull().sum()

KeyError: "['wind_from_direction'] not in index"

In [133]:
# Same with wind speed
(sarasota["wind_speed"].describe())
negative_speed_count = (sarasota["wind_speed"] < 0).sum()
(f"Negative wind speed readings: {negative_speed_count}")

# Wind speed can never be negative
if negative_speed_count > 0:
    sarasota.loc[sarasota["wind_speed"] < 0, "wind_speed"] = np.nan
    sarasota["wind_speed"] = sarasota["wind_speed"].interpolate()
    ("Negative wind speeds set to NaN and interpolated.")
sarasota[sarasota["wind_speed"] > 20][
    ["time", "wind_speed", "wind_from_direction"]
]

KeyError: "['wind_from_direction'] not in index"

In [136]:
os.chdir(DATA_DIR+"/cleaned")
bradentonFL.to_csv("bradentonFL_tempData_cleaned.csv", index=False)
print("Saved cleaned dataset to bradentonFL_tempData_cleaned.csv")
floridaCyanotoxins.to_csv("florida_cyanotoxins_2015-01-01_to_2026-06-17_cleaned.csv", index=False)
print("Saved cleaned dataset to florida_cyanotoxins_2015-01-01_to_2026-06-17_cleaned.csv")
habsos.to_csv("habsos_cellcounts_cleaned.csv", index=False)
print("Saved cleaned dataset to habsos_cellcounts_cleaned.csv")
motemarine.to_csv("MoteMarine_BottomTemp_cleaned.csv", index=False)
print("Saved cleaned dataset to MoteMarine_BottomTemp_cleaned.csv")
nutrientsFL.to_csv("NutrientsFL_2006_2025_cleaned.csv", index=False)
print("Saved cleaned dataset to NutrientsFL_2006_2025_cleaned.csv")
# sarasota.to_csv("Sarasota_Wind&Temp.csv", index=False)
# print("Saved cleaned dataset to Sarasota_Wind&Temp.csv")


Saved cleaned dataset to bradentonFL_tempData_cleaned.csv
Saved cleaned dataset to florida_cyanotoxins_2015-01-01_to_2026-06-17_cleaned.csv
Saved cleaned dataset to habsos_cellcounts_cleaned.csv
Saved cleaned dataset to MoteMarine_BottomTemp_cleaned.csv
Saved cleaned dataset to NutrientsFL_2006_2025_cleaned.csv
